---
title: "immuneATAC_SRA"
author: "Alex Chitsazan"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: html_document
---
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## SNAPATAC



```{r}


## SNAPATAC
library(SnapATAC)
library(leiden)
# ## List of files 
# file.list <- list.files(path = "~/Documents/snapMirror", pattern = ".*.snap", recursive =F, full.names = T)
# 
# 
# sample.list <- c( "Weird_sample_G4", "G33_2", "G33_1", "G4_1")


###################
#####  G33_1  #####
###################
G33_1.sp = createSnap(
	file="Ece_G33.snap",
	sample="G33_1",
	num.cores=11
	)
### G33_1 prefiltering
plotBarcode(
  obj=G33_1.sp, 
	pdf.file.name=NULL, 
	pdf.width=7, 
	pdf.height=7, 
	col="grey",
	border="grey",
	breaks=50
	) 
# Filter the plots
G33_1.sp = filterCells(
	obj=G33_1.sp,
	subset.names=c("fragment.num", "UMI"),
	low.thresholds=c(1000,500),
	high.thresholds=c(Inf, Inf)
	)



## add bMAT
G33_1.sp = addBmatToSnap(G33_1.sp, bin.size=5000, num.cores=11)
G33_1.sp = makeBinary(G33_1.sp, mat="bmat");


### Genes.gr and promoters.gr
library(GenomicRanges)
genes <- read.table("gencode_v19_gene_pos.bed");
genes.gr = GRanges(genes[,1],
                   IRanges(genes[,2], genes[,3]),
                   name=genes[,4]
)
promoter.gr <- promoters(genes.gr)
ov = findOverlaps(G33_1.sp@feature, promoter.gr);
idy = queryHits(ov);
promoter_ratio = SnapATAC::rowSums(G33_1.sp[,idy, mat="bmat"], mat="bmat") / SnapATAC::rowSums(G33_1.sp, mat="bmat");
plot(log(SnapATAC::rowSums(G33_1.sp, mat="bmat") + 1,10), promoter_ratio, cex=0.5, col="grey", xlab="log(UMI)", ylab="FIP Ratio", ylim=c(0,1 ))
abline(h = 0.2, col="red")
idx = which(promoter_ratio > 0.2 & promoter_ratio < 0.8);
G33_1.sp = G33_1.sp[idx,]
G33_1.sp@metaData$MTRatio <- (G33_1.sp@metaData$CM+1) / (G33_1.sp@metaData$UQ+1)
G33_1.sp@metaData$PromoterRatio <- promoter_ratio[idx]
plot(G33_1.sp@metaData$PromoterRatio, G33_1.sp@metaData$MTRatio, cex=0.5, xlim=c(0,1), main = "G33_1") 

## Filter using Andrew's method
# barcodes <- read.table("~/Box Sync/Alex_Data/Ece_scATAC/snapATAC/Data/COMBINED.counts.filt_100_10.cistopic.G3_12_G4_Immune.annot")
# trueBarcodes <- match(barcodes[,1], G33_1.sp@barcode)
# table(is.na(trueBarcodes))
# trueBarcodes <- na.omit(trueBarcodes)
# G33_1.sp <- G33_1.sp[trueBarcodes,, mat="bmat"]


## After Filteration
summarySnap(G33_1.sp)
plotBarcode(
  obj=G33_1.sp, 
	pdf.file.name=NULL, 
	pdf.width=7, 
	pdf.height=7, 
	col="grey",
	border="grey",
	breaks=50
	)



###################
#####  G33_2  #####
###################

G33_2.sp = createSnap(
	file="Ece_G33_2.snap",
	sample="G33_2",
	num.cores=11
	)
### G33_2 prefiltering
plotBarcode(
  obj=G33_2.sp,
	pdf.file.name=NULL,
	pdf.width=7,
	pdf.height=7,
	col="grey",
	border="grey",
	breaks=50
	)
# trueBarcodes <- match(barcodes[,1], G33_2.sp@barcode)
# trueBarcodes <- na.omit(trueBarcodes)
# G33_2.sp <- G33_2.sp[trueBarcodes,, mat="bmat"]

## Filter Cells
G33_2.sp = filterCells(
  obj=G33_2.sp,
  subset.names=c("fragment.num", "UMI"),
  low.thresholds=c(1000,500),
  high.thresholds=c(Inf, Inf)
)
## add bMAT
showBinSizes("Ece_G33_2.snap") 
G33_2.sp = addBmatToSnap(G33_2.sp, bin.size=5000, num.cores=11)
G33_2.sp = makeBinary(G33_2.sp, mat="bmat");

ov = findOverlaps(G33_2.sp@feature, promoter.gr);
idy = queryHits(ov);
promoter_ratio = SnapATAC::rowSums(G33_2.sp[,idy, mat="bmat"], mat="bmat") / SnapATAC::rowSums(G33_2.sp, mat="bmat");
plot(log(SnapATAC::rowSums(G33_2.sp, mat="bmat") + 1,10), promoter_ratio, cex=0.5, col="grey", xlab="log(UMI)", ylab="FIP Ratio", ylim=c(0,1 ))
abline(h = 0.2, col="red")
idx = which(promoter_ratio > 0.2 & promoter_ratio < 0.8);
G33_2.sp = G33_2.sp[idx,]
G33_2.sp@metaData$MTRatio <- (G33_2.sp@metaData$CM+1) / (G33_2.sp@metaData$UQ+1)
G33_2.sp@metaData$PromoterRatio <- promoter_ratio[idx]
plot(G33_2.sp@metaData$PromoterRatio, G33_2.sp@metaData$MTRatio, cex=0.5, xlim=c(0,1), main = "G33_2") 

###################
#####  G4_1  #####
###################

G4_1.sp = createSnap(
	file = "Ece_G4_4.snap",
	sample="G4_1",
	num.cores=11
	)
### G4_1 prefiltering
plotBarcode(
  obj=G4_1.sp, 
	pdf.file.name=NULL, 
	pdf.width=7, 
	pdf.height=7, 
	col="grey",
	border="grey",
	breaks=50
	)

# Filter Cells
G4_1.sp = filterCells(
  obj=G4_1.sp,
  subset.names=c("fragment.num", "UMI"),
  low.thresholds=c(1000,500),
  high.thresholds=c(Inf, Inf)
	)
# ## Filter based upon previous filter
# trueBarcodes <- match(barcodes[,1], G4_1.sp@barcode)
# trueBarcodes <- na.omit(trueBarcodes)
# G4_1.sp <- G4_1.sp[trueBarcodes,, mat="bmat"]

## add bMAT
showBinSizes("Ece_G4_4.snap")
G4_1.sp = addBmatToSnap(G4_1.sp, bin.size=5000, num.cores=11)
G4_1.sp = makeBinary(G4_1.sp, mat="bmat");

### Genes.gr and promoters.gr
ov = findOverlaps(G4_1.sp@feature, promoter.gr);
idy = queryHits(ov);

promoter_ratio = SnapATAC::rowSums(G4_1.sp[,idy, mat="bmat"], mat="bmat") / SnapATAC::rowSums(G4_1.sp, mat="bmat");
plot(log(SnapATAC::rowSums(G4_1.sp, mat="bmat") + 1,10), promoter_ratio, cex=0.5, col="grey", xlab="log(UMI)", ylab="FIP Ratio", ylim=c(0,1 ));
idx = which(promoter_ratio > 0.2 & promoter_ratio < 0.8);
G4_1.sp = G4_1.sp[idx,]
G4_1.sp@metaData$MTRatio <- (G4_1.sp@metaData$CM+1) / (G4_1.sp@metaData$UQ+1)
G4_1.sp@metaData$PromoterRatio <- promoter_ratio[idx]

## After filteration
summarySnap(G4_1.sp)
plotBarcode(
  obj=G4_1.sp, 
	pdf.file.name=NULL, 
	pdf.width=7, 
	pdf.height=7, 
	col="grey",
	border="grey",
	breaks=50
	)








### new 

########################
#####    #####
########################


sample.list <- c("190701", "190717")
Sample_190701.sp = createSnap(
	file="190701_NS500556_0312_AH2N2JBGXB.ECE.snap",
	sample="G33_3",
	num.cores=11
	)
### Weird prefiltering
plotBarcode(
  obj=Sample_190701.sp, 
	pdf.file.name=NULL, 
	pdf.width=7, 
	pdf.height=7, 
	col="grey",
	border="grey",
	breaks=50
	)

# Filter Cells
Sample_190701.sp = filterCells(
	obj=Sample_190701.sp,
	subset.names=c("fragment.num", "UMI"),
	low.thresholds=c(1000,700),
	high.thresholds=c(Inf, Inf)
	)

# ## Filter based upon previous filter
# trueBarcodes <- match(barcodes[,1], Sample_190701.sp@barcode)
# trueBarcodes <- na.omit(trueBarcodes)
# Sample_190701.sp <- Sample_190701.sp[trueBarcodes,, mat="bmat"]

## add bMAT
Sample_190701.sp = addBmatToSnap(Sample_190701.sp, bin.size=5000, num.cores=11)
Sample_190701.sp = makeBinary(Sample_190701.sp, mat="bmat");
table(seqnames(AllSamplesDecember.lda.sp@feature))
findOverlaps(genes.sel.gr["AR"], AllSamplesDecember.lda.sp@feature)
### Genes.gr and promoters.gr
ov = findOverlaps(Sample_190701.sp@feature, promoter.gr);
idy = queryHits(ov);

promoter_ratio = SnapATAC::rowSums(Sample_190701.sp[,idy, mat="bmat"], mat="bmat") / SnapATAC::rowSums(Sample_190701.sp, mat="bmat");
plot(log(SnapATAC::rowSums(Sample_190701.sp, mat="bmat") + 1,10), promoter_ratio, cex=0.5, col="grey", xlab="log(UMI)", ylab="FIP Ratio", ylim=c(0,1 ));
idx = which(promoter_ratio > 0.2 & promoter_ratio < 0.8);
Sample_190701.sp = Sample_190701.sp[idx,]
Sample_190701.sp@metaData$MTRatio <- (Sample_190701.sp@metaData$CM+1) / (Sample_190701.sp@metaData$UQ+1)
Sample_190701.sp@metaData$PromoterRatio <- promoter_ratio[idx]

## After filteration
summarySnap(Sample_190701.sp)
plotBarcode(
  obj=Sample_190701.sp, 
	pdf.file.name=NULL, 
	pdf.width=7, 
	pdf.height=7, 
	col="grey",
	border="grey",
	breaks=50
	)






G4_2.sp = createSnap(
	file="ID6683.snap",
	sample="G4_2",
	num.cores=11
	)
### G4_2 prefiltering
plotBarcode(
  obj=G4_2.sp,
	pdf.file.name=NULL,
	pdf.width=7,
	pdf.height=7,
	col="grey",
	border="grey",
	breaks=50
	)
# trueBarcodes <- match(barcodes[,1], G4_2.sp@barcode)
# trueBarcodes <- na.omit(trueBarcodes)
# G4_2.sp <- G4_2.sp[trueBarcodes,, mat="bmat"]

## Filter Cells
G4_2.sp = filterCells(
  obj=G4_2.sp,
  subset.names=c("fragment.num", "UMI"),
  low.thresholds=c(1000,1000),
  high.thresholds=c(Inf, Inf)
)
## add bMAT
G4_2.sp = addBmatToSnap(G4_2.sp, bin.size=5000, num.cores=11)
G4_2.sp = makeBinary(G4_2.sp, mat="bmat");

ov = findOverlaps(G4_2.sp@feature, promoter.gr);
idy = queryHits(ov);
promoter_ratio = SnapATAC::rowSums(G4_2.sp[,idy, mat="bmat"], mat="bmat") / SnapATAC::rowSums(G4_2.sp, mat="bmat");
plot(log(SnapATAC::rowSums(G4_2.sp, mat="bmat") + 1,10), promoter_ratio, cex=0.5, col="grey", xlab="log(UMI)", ylab="FIP Ratio", ylim=c(0,1 ))
abline(h = 0.2, col="red")
idx = which(promoter_ratio > 0.2 & promoter_ratio < 0.8);
G4_2.sp = G4_2.sp[idx,]
G4_2.sp@metaData$MTRatio <- (G4_2.sp@metaData$CM+1) / (G4_2.sp@metaData$UQ+1)
G4_2.sp@metaData$PromoterRatio <- promoter_ratio[idx]
plot(G4_2.sp@metaData$PromoterRatio, G4_2.sp@metaData$MTRatio, cex=0.5, xlim=c(0,1), main = "G4_2") 

## After Filteration
summarySnap(G4_2.sp)
plotBarcode(
  obj=G4_2.sp, 
	pdf.file.name=NULL, 
	pdf.width=7, 
	pdf.height=7, 
	col="grey",
	border="grey",
	breaks=50
	)



###################
#####  G4_1  #####
###################

TB12146.sp = createSnap(
	file = "TB12146.snap",
	sample="TB12146",
	num.cores=11
	)
### TB12146 prefiltering
plotBarcode(
  obj=TB12146.sp, 
	pdf.file.name=NULL, 
	pdf.width=7, 
	pdf.height=7, 
	col="grey",
	border="grey",
	breaks=50
	)

# Filter Cells
TB12146.sp = filterCells(
  obj=TB12146.sp,
  subset.names=c("fragment.num", "UMI"),
  low.thresholds=c(1000,500),
  high.thresholds=c(Inf, Inf)
	)
# ## Filter based upon previous filter
# trueBarcodes <- match(barcodes[,1], TB12146.sp@barcode)
# trueBarcodes <- na.omit(trueBarcodes)
# TB12146.sp <- TB12146.sp[trueBarcodes,, mat="bmat"]

## add bMAT
TB12146.sp = addBmatToSnap(TB12146.sp, bin.size=5000, num.cores=11)
TB12146.sp = makeBinary(TB12146.sp, mat="bmat");

### Genes.gr and promoters.gr
ov = findOverlaps(TB12146.sp@feature, promoter.gr);
idy = queryHits(ov);

promoter_ratio = SnapATAC::rowSums(TB12146.sp[,idy, mat="bmat"], mat="bmat") / SnapATAC::rowSums(TB12146.sp, mat="bmat");
plot(log(SnapATAC::rowSums(TB12146.sp, mat="bmat") + 1,10), promoter_ratio, cex=0.5, col="grey", xlab="log(UMI)", ylab="FIP Ratio", ylim=c(0,1 ));
idx = which(promoter_ratio > 0.2 & promoter_ratio < 0.8);
TB12146.sp = TB12146.sp[idx,]
TB12146.sp@metaData$MTRatio <- (TB12146.sp@metaData$CM+1) / (TB12146.sp@metaData$UQ+1)
TB12146.sp@metaData$PromoterRatio <- promoter_ratio[idx]

## After filteration
summarySnap(TB12146.sp)
plotBarcode(
  obj=TB12146.sp, 
	pdf.file.name=NULL, 
	pdf.width=7, 
	pdf.height=7, 
	col="grey",
	border="grey",
	breaks=50
	)









TB4732.sp = createSnap(
	file="TB4732.snap",
	sample="TB4732",
	num.cores=11
	)
### Weird prefiltering
plotBarcode(
  obj=TB4732.sp, 
	pdf.file.name=NULL, 
	pdf.width=7, 
	pdf.height=7, 
	col="grey",
	border="grey",
	breaks=50
	)

# Filter Cells
TB4732.sp = filterCells(
	obj=TB4732.sp,
	subset.names=c("fragment.num", "UMI"),
	low.thresholds=c(1000,500),
	high.thresholds=c(Inf, Inf)
	)

# ## Filter based upon previous filter
# trueBarcodes <- match(barcodes[,1], TB4732.sp@barcode)
# trueBarcodes <- na.omit(trueBarcodes)
# TB4732.sp <- TB4732.sp[trueBarcodes,, mat="bmat"]

## add bMAT
TB4732.sp = addBmatToSnap(TB4732.sp, bin.size=5000, num.cores=11)
TB4732.sp = makeBinary(TB4732.sp, mat="bmat");

### Genes.gr and promoters.gr
ov = findOverlaps(TB4732.sp@feature, promoter.gr);
idy = queryHits(ov);

promoter_ratio = SnapATAC::rowSums(TB4732.sp[,idy, mat="bmat"], mat="bmat") / SnapATAC::rowSums(TB4732.sp, mat="bmat");
plot(log(SnapATAC::rowSums(TB4732.sp, mat="bmat") + 1,10), promoter_ratio, cex=0.5, col="grey", xlab="log(UMI)", ylab="FIP Ratio", ylim=c(0,1 ));
idx = which(promoter_ratio > 0.2 & promoter_ratio < 0.8);
TB4732.sp = TB4732.sp[idx,]
TB4732.sp@metaData$MTRatio <- (TB4732.sp@metaData$CM+1) / (TB4732.sp@metaData$UQ+1)
TB4732.sp@metaData$PromoterRatio <- promoter_ratio[idx]

## After filteration
summarySnap(TB4732.sp)
plotBarcode(
  obj=TB4732.sp, 
	pdf.file.name=NULL, 
	pdf.width=7, 
	pdf.height=7, 
	col="grey",
	border="grey",
	breaks=50
	)







TB5020.sp = createSnap(
	file="TB5020.snap",
	sample="TB5020",
	num.cores=11
	)
### Weird prefiltering
plotBarcode(
  obj=TB5020.sp, 
	pdf.file.name=NULL, 
	pdf.width=7, 
	pdf.height=7, 
	col="grey",
	border="grey",
	breaks=50
	)

# Filter Cells
TB5020.sp = filterCells(
	obj=TB5020.sp,
	subset.names=c("fragment.num", "UMI"),
	low.thresholds=c(1000,500),
	high.thresholds=c(Inf, Inf)
	)

# ## Filter based upon previous filter
# trueBarcodes <- match(barcodes[,1], TB5020.sp@barcode)
# trueBarcodes <- na.omit(trueBarcodes)
# TB5020.sp <- TB5020.sp[trueBarcodes,, mat="bmat"]

## add bMAT
TB5020.sp = addBmatToSnap(TB5020.sp, bin.size=5000, num.cores=11)
TB5020.sp = makeBinary(TB5020.sp, mat="bmat");

### Genes.gr and promoters.gr
ov = findOverlaps(TB5020.sp@feature, promoter.gr);
idy = queryHits(ov);

promoter_ratio = SnapATAC::rowSums(TB5020.sp[,idy, mat="bmat"], mat="bmat") / SnapATAC::rowSums(TB5020.sp, mat="bmat");
plot(log(SnapATAC::rowSums(TB5020.sp, mat="bmat") + 1,10), promoter_ratio, cex=0.5, col="grey", xlab="log(UMI)", ylab="FIP Ratio", ylim=c(0,1 ));
idx = which(promoter_ratio > 0.2 & promoter_ratio < 0.8);
TB5020.sp = TB5020.sp[idx,]
TB5020.sp@metaData$MTRatio <- (TB5020.sp@metaData$CM+1) / (TB5020.sp@metaData$UQ+1)
TB5020.sp@metaData$PromoterRatio <- promoter_ratio[idx]

## After filteration
summarySnap(TB5020.sp)
plotBarcode(
  obj=TB5020.sp, 
	pdf.file.name=NULL, 
	pdf.width=7, 
	pdf.height=7, 
	col="grey",
	border="grey",
	breaks=50
	)








TB6475.sp = createSnap(
	file="TB6475.snap",
	sample="TB6475",
	num.cores=11
	)
### Weird prefiltering
plotBarcode(
  obj=TB6475.sp, 
	pdf.file.name=NULL, 
	pdf.width=7, 
	pdf.height=7, 
	col="grey",
	border="grey",
	breaks=50
	)

# Filter Cells
TB6475.sp = filterCells(
	obj=TB6475.sp,
	subset.names=c("fragment.num", "UMI"),
	low.thresholds=c(1000,500),
	high.thresholds=c(Inf, Inf)
	)

# ## Filter based upon previous filter
# trueBarcodes <- match(barcodes[,1], TB6475.sp@barcode)
# trueBarcodes <- na.omit(trueBarcodes)
# TB6475.sp <- TB6475.sp[trueBarcodes,, mat="bmat"]

## add bMAT
TB6475.sp = addBmatToSnap(TB6475.sp, bin.size=5000, num.cores=11)
TB6475.sp = makeBinary(TB6475.sp, mat="bmat");

### Genes.gr and promoters.gr
ov = findOverlaps(TB6475.sp@feature, promoter.gr);
idy = queryHits(ov);

promoter_ratio = SnapATAC::rowSums(TB6475.sp[,idy, mat="bmat"], mat="bmat") / SnapATAC::rowSums(TB6475.sp, mat="bmat");
plot(log(SnapATAC::rowSums(TB6475.sp, mat="bmat") + 1,10), promoter_ratio, cex=0.5, col="grey", xlab="log(UMI)", ylab="FIP Ratio", ylim=c(0,1 ));
idx = which(promoter_ratio > 0.2 & promoter_ratio < 0.8);
TB6475.sp = TB6475.sp[idx,]
TB6475.sp@metaData$MTRatio <- (TB6475.sp@metaData$CM+1) / (TB6475.sp@metaData$UQ+1)
TB6475.sp@metaData$PromoterRatio <- promoter_ratio[idx]

## After filteration
summarySnap(TB6475.sp)
plotBarcode(
  obj=TB6475.sp, 
	pdf.file.name=NULL, 
	pdf.width=7, 
	pdf.height=7, 
	col="grey",
	border="grey",
	breaks=50
	)




TB7147.sp = createSnap(
	file="TB7147.snap",
	sample="TB7147",
	num.cores=11
	)
### Weird prefiltering
plotBarcode(
  obj=TB7147.sp, 
	pdf.file.name=NULL, 
	pdf.width=7, 
	pdf.height=7, 
	col="grey",
	border="grey",
	breaks=50
	)

# Filter Cells
TB7147.sp = filterCells(
	obj=TB7147.sp,
	subset.names=c("fragment.num", "UMI"),
	low.thresholds=c(1000,500),
	high.thresholds=c(Inf, Inf)
	)

# ## Filter based upon previous filter
# trueBarcodes <- match(barcodes[,1], TB7147.sp@barcode)
# trueBarcodes <- na.omit(trueBarcodes)
# TB7147.sp <- TB7147.sp[trueBarcodes,, mat="bmat"]

## add bMAT
TB7147.sp = addBmatToSnap(TB7147.sp, bin.size=5000, num.cores=11)
TB7147.sp = makeBinary(TB7147.sp, mat="bmat");

### Genes.gr and promoters.gr
ov = findOverlaps(TB7147.sp@feature, promoter.gr);
idy = queryHits(ov);

promoter_ratio = SnapATAC::rowSums(TB7147.sp[,idy, mat="bmat"], mat="bmat") / SnapATAC::rowSums(TB7147.sp, mat="bmat");
plot(log(SnapATAC::rowSums(TB7147.sp, mat="bmat") + 1,10), promoter_ratio, cex=0.5, col="grey", xlab="log(UMI)", ylab="FIP Ratio", ylim=c(0,1 ));
idx = which(promoter_ratio > 0.2 & promoter_ratio < 0.8);
TB7147.sp = TB7147.sp[idx,]
TB7147.sp@metaData$MTRatio <- (TB7147.sp@metaData$CM+1) / (TB7147.sp@metaData$UQ+1)
TB7147.sp@metaData$PromoterRatio <- promoter_ratio[idx]

## After filteration
summarySnap(TB7147.sp)
plotBarcode(
  obj=TB7147.sp, 
	pdf.file.name=NULL, 
	pdf.width=7, 
	pdf.height=7, 
	col="grey",
	border="grey",
	breaks=50
	)








TB7521.sp = createSnap(
	file="TB7521.snap",
	sample="TB7521",
	num.cores=11
	)
### Weird prefiltering
plotBarcode(
  obj=TB7521.sp, 
	pdf.file.name=NULL, 
	pdf.width=7, 
	pdf.height=7, 
	col="grey",
	border="grey",
	breaks=50
	)

# Filter Cells
TB7521.sp = filterCells(
	obj=TB7521.sp,
	subset.names=c("fragment.num", "UMI"),
	low.thresholds=c(1000,500),
	high.thresholds=c(Inf, Inf)
	)

# ## Filter based upon previous filter
# trueBarcodes <- match(barcodes[,1], TB7521.sp@barcode)
# trueBarcodes <- na.omit(trueBarcodes)
# TB7521.sp <- TB7521.sp[trueBarcodes,, mat="bmat"]

## add bMAT
TB7521.sp = addBmatToSnap(TB7521.sp, bin.size=5000, num.cores=11)
TB7521.sp = makeBinary(TB7521.sp, mat="bmat");

### Genes.gr and promoters.gr
ov = findOverlaps(TB7521.sp@feature, promoter.gr);
idy = queryHits(ov);

promoter_ratio = SnapATAC::rowSums(TB7521.sp[,idy, mat="bmat"], mat="bmat") / SnapATAC::rowSums(TB7521.sp, mat="bmat");
plot(log(SnapATAC::rowSums(TB7521.sp, mat="bmat") + 1,10), promoter_ratio, cex=0.5, col="grey", xlab="log(UMI)", ylab="FIP Ratio", ylim=c(0,1 ));
idx = which(promoter_ratio > 0.2 & promoter_ratio < 0.8);
TB7521.sp = TB7521.sp[idx,]
TB7521.sp@metaData$MTRatio <- (TB7521.sp@metaData$CM+1) / (TB7521.sp@metaData$UQ+1)
TB7521.sp@metaData$PromoterRatio <- promoter_ratio[idx]

## After filteration
summarySnap(TB7521.sp)
plotBarcode(
  obj=TB7521.sp, 
	pdf.file.name=NULL, 
	pdf.width=7, 
	pdf.height=7, 
	col="grey",
	border="grey",
	breaks=50
	)











TB7677.sp = createSnap(
	file="TB7677.snap",
	sample="TB7677",
	num.cores=11
	)
### Weird prefiltering
plotBarcode(
  obj=TB7677.sp, 
	pdf.file.name=NULL, 
	pdf.width=7, 
	pdf.height=7, 
	col="grey",
	border="grey",
	breaks=50
	)

# Filter Cells
TB7677.sp = filterCells(
	obj=TB7677.sp,
	subset.names=c("fragment.num", "UMI"),
	low.thresholds=c(1000,500),
	high.thresholds=c(Inf, Inf)
	)

# ## Filter based upon previous filter
# trueBarcodes <- match(barcodes[,1], TB7677.sp@barcode)
# trueBarcodes <- na.omit(trueBarcodes)
# TB7677.sp <- TB7677.sp[trueBarcodes,, mat="bmat"]

## add bMAT
TB7677.sp = addBmatToSnap(TB7677.sp, bin.size=5000, num.cores=11)
TB7677.sp = makeBinary(TB7677.sp, mat="bmat");

### Genes.gr and promoters.gr
ov = findOverlaps(TB7677.sp@feature, promoter.gr);
idy = queryHits(ov);

promoter_ratio = SnapATAC::rowSums(TB7677.sp[,idy, mat="bmat"], mat="bmat") / SnapATAC::rowSums(TB7677.sp, mat="bmat");
plot(log(SnapATAC::rowSums(TB7677.sp, mat="bmat") + 1,10), promoter_ratio, cex=0.5, col="grey", xlab="log(UMI)", ylab="FIP Ratio", ylim=c(0,1 ));
idx = which(promoter_ratio > 0.2 & promoter_ratio < 0.8);
TB7677.sp = TB7677.sp[idx,]
TB7677.sp@metaData$MTRatio <- (TB7677.sp@metaData$CM+1) / (TB7677.sp@metaData$UQ+1)
TB7677.sp@metaData$PromoterRatio <- promoter_ratio[idx]

## After filteration
summarySnap(TB7677.sp)
plotBarcode(
  obj=TB7677.sp, 
	pdf.file.name=NULL, 
	pdf.width=7, 
	pdf.height=7, 
	col="grey",
	border="grey",
	breaks=50
	)














TB7799.sp = createSnap(
	file="TB7799.snap",
	sample="TB7799",
	num.cores=11
	)
### Weird prefiltering
plotBarcode(
  obj=TB7799.sp, 
	pdf.file.name=NULL, 
	pdf.width=7, 
	pdf.height=7, 
	col="grey",
	border="grey",
	breaks=50
	)

# Filter Cells
TB7799.sp = filterCells(
	obj=TB7799.sp,
	subset.names=c("fragment.num", "UMI"),
	low.thresholds=c(1000,500),
	high.thresholds=c(Inf, Inf)
	)

# ## Filter based upon previous filter
# trueBarcodes <- match(barcodes[,1], TB7799.sp@barcode)
# trueBarcodes <- na.omit(trueBarcodes)
# TB7799.sp <- TB7799.sp[trueBarcodes,, mat="bmat"]

## add bMAT
TB7799.sp = addBmatToSnap(TB7799.sp, bin.size=5000, num.cores=11)
TB7799.sp = makeBinary(TB7799.sp, mat="bmat");

### Genes.gr and promoters.gr
ov = findOverlaps(TB7799.sp@feature, promoter.gr);
idy = queryHits(ov);

promoter_ratio = SnapATAC::rowSums(TB7799.sp[,idy, mat="bmat"], mat="bmat") / SnapATAC::rowSums(TB7799.sp, mat="bmat");
plot(log(SnapATAC::rowSums(TB7799.sp, mat="bmat") + 1,10), promoter_ratio, cex=0.5, col="grey", xlab="log(UMI)", ylab="FIP Ratio", ylim=c(0,1 ));
idx = which(promoter_ratio > 0.2 & promoter_ratio < 0.8);
TB7799.sp = TB7799.sp[idx,]
TB7799.sp@metaData$MTRatio <- (TB7799.sp@metaData$CM+1) / (TB7799.sp@metaData$UQ+1)
TB7799.sp@metaData$PromoterRatio <- promoter_ratio[idx]

## After filteration
summarySnap(TB7799.sp)
plotBarcode(
  obj=TB7799.sp, 
	pdf.file.name=NULL, 
	pdf.width=7, 
	pdf.height=7, 
	col="grey",
	border="grey",
	breaks=50
	)













TB9679.sp = createSnap(
	file="TB9679.snap",
	sample="TB9679",
	num.cores=11
	)
### Weird prefiltering
plotBarcode(
  obj=TB9679.sp, 
	pdf.file.name=NULL, 
	pdf.width=7, 
	pdf.height=7, 
	col="grey",
	border="grey",
	breaks=50
	)

# Filter Cells
TB9679.sp = filterCells(
	obj=TB9679.sp,
	subset.names=c("fragment.num", "UMI"),
	low.thresholds=c(1000,500),
	high.thresholds=c(Inf, Inf)
	)

# ## Filter based upon previous filter
# trueBarcodes <- match(barcodes[,1], TB9679.sp@barcode)
# trueBarcodes <- na.omit(trueBarcodes)
# TB9679.sp <- TB9679.sp[trueBarcodes,, mat="bmat"]

## add bMAT
TB9679.sp = addBmatToSnap(TB9679.sp, bin.size=5000, num.cores=11)
TB9679.sp = makeBinary(TB9679.sp, mat="bmat");

### Genes.gr and promoters.gr
ov = findOverlaps(TB9679.sp@feature, promoter.gr);
idy = queryHits(ov);

promoter_ratio = SnapATAC::rowSums(TB9679.sp[,idy, mat="bmat"], mat="bmat") / SnapATAC::rowSums(TB9679.sp, mat="bmat");
plot(log(SnapATAC::rowSums(TB9679.sp, mat="bmat") + 1,10), promoter_ratio, cex=0.5, col="grey", xlab="log(UMI)", ylab="FIP Ratio", ylim=c(0,1 ));
idx = which(promoter_ratio > 0.2 & promoter_ratio < 0.8);
TB9679.sp = TB9679.sp[idx,]
TB9679.sp@metaData$MTRatio <- (TB9679.sp@metaData$CM+1) / (TB9679.sp@metaData$UQ+1)
TB9679.sp@metaData$PromoterRatio <- promoter_ratio[idx]

## After filteration
summarySnap(TB9679.sp)
plotBarcode(
  obj=TB9679.sp, 
	pdf.file.name=NULL, 
	pdf.width=7, 
	pdf.height=7, 
	col="grey",
	border="grey",
	breaks=50
	)











WM306009.sp = createSnap(
	file="WM306009.snap",
	sample="WM306009",
	num.cores=11
	)
### Weird prefiltering
plotBarcode(
  obj=WM306009.sp, 
	pdf.file.name=NULL, 
	pdf.width=7, 
	pdf.height=7, 
	col="grey",
	border="grey",
	breaks=50
	)

# Filter Cells
WM306009.sp = filterCells(
	obj=WM306009.sp,
	subset.names=c("fragment.num", "UMI"),
	low.thresholds=c(1000,500),
	high.thresholds=c(Inf, Inf)
	)

# ## Filter based upon previous filter
# trueBarcodes <- match(barcodes[,1], WM306009.sp@barcode)
# trueBarcodes <- na.omit(trueBarcodes)
# WM306009.sp <- WM306009.sp[trueBarcodes,, mat="bmat"]

## add bMAT
WM306009.sp = addBmatToSnap(WM306009.sp, bin.size=5000, num.cores=11)
WM306009.sp = makeBinary(WM306009.sp, mat="bmat");

### Genes.gr and promoters.gr
ov = findOverlaps(WM306009.sp@feature, promoter.gr);
idy = queryHits(ov);

promoter_ratio = SnapATAC::rowSums(WM306009.sp[,idy, mat="bmat"], mat="bmat") / SnapATAC::rowSums(WM306009.sp, mat="bmat");
plot(log(SnapATAC::rowSums(WM306009.sp, mat="bmat") + 1,10), promoter_ratio, cex=0.5, col="grey", xlab="log(UMI)", ylab="FIP Ratio", ylim=c(0,1 ));
idx = which(promoter_ratio > 0.2 & promoter_ratio < 0.8);
WM306009.sp = WM306009.sp[idx,]
WM306009.sp@metaData$MTRatio <- (WM306009.sp@metaData$CM+1) / (WM306009.sp@metaData$UQ+1)
WM306009.sp@metaData$PromoterRatio <- promoter_ratio[idx]

## After filteration
summarySnap(WM306009.sp)
plotBarcode(
  obj=WM306009.sp, 
	pdf.file.name=NULL, 
	pdf.width=7, 
	pdf.height=7, 
	col="grey",
	border="grey",
	breaks=50
	)












WM306010.sp = createSnap(
	file="WM306010.snap",
	sample="WM306010",
	num.cores=11
	)
### Weird prefiltering
plotBarcode(
  obj=WM306010.sp, 
	pdf.file.name=NULL, 
	pdf.width=7, 
	pdf.height=7, 
	col="grey",
	border="grey",
	breaks=50
	)

# Filter Cells
WM306010.sp = filterCells(
	obj=WM306010.sp,
	subset.names=c("fragment.num", "UMI"),
	low.thresholds=c(1000,500),
	high.thresholds=c(Inf, Inf)
	)

# ## Filter based upon previous filter
# trueBarcodes <- match(barcodes[,1], WM306010.sp@barcode)
# trueBarcodes <- na.omit(trueBarcodes)
# WM306010.sp <- WM306010.sp[trueBarcodes,, mat="bmat"]

## add bMAT
WM306010.sp = addBmatToSnap(WM306010.sp, bin.size=5000, num.cores=11)
WM306010.sp = makeBinary(WM306010.sp, mat="bmat");

### Genes.gr and promoters.gr
ov = findOverlaps(WM306010.sp@feature, promoter.gr);
idy = queryHits(ov);

promoter_ratio = SnapATAC::rowSums(WM306010.sp[,idy, mat="bmat"], mat="bmat") / SnapATAC::rowSums(WM306010.sp, mat="bmat");
plot(log(SnapATAC::rowSums(WM306010.sp, mat="bmat") + 1,10), promoter_ratio, cex=0.5, col="grey", xlab="log(UMI)", ylab="FIP Ratio", ylim=c(0,1 ));
idx = which(promoter_ratio > 0.2 & promoter_ratio < 0.8);
WM306010.sp = WM306010.sp[idx,]
WM306010.sp@metaData$MTRatio <- (WM306010.sp@metaData$CM+1) / (WM306010.sp@metaData$UQ+1)
WM306010.sp@metaData$PromoterRatio <- promoter_ratio[idx]

## After filteration
summarySnap(WM306010.sp)
plotBarcode(
  obj=WM306010.sp, 
	pdf.file.name=NULL, 
	pdf.width=7, 
	pdf.height=7, 
	col="grey",
	border="grey",
	breaks=50
	)
















WM306015.sp = createSnap(
	file="WM306015.snap",
	sample="WM306015",
	num.cores=11
	)
### Weird prefiltering
plotBarcode(
  obj=WM306015.sp, 
	pdf.file.name=NULL, 
	pdf.width=7, 
	pdf.height=7, 
	col="grey",
	border="grey",
	breaks=50
	)

# Filter Cells
WM306015.sp = filterCells(
	obj=WM306015.sp,
	subset.names=c("fragment.num", "UMI"),
	low.thresholds=c(1000,500),
	high.thresholds=c(Inf, Inf)
	)

# ## Filter based upon previous filter
# trueBarcodes <- match(barcodes[,1], WM306015.sp@barcode)
# trueBarcodes <- na.omit(trueBarcodes)
# WM306015.sp <- WM306015.sp[trueBarcodes,, mat="bmat"]

## add bMAT
WM306015.sp = addBmatToSnap(WM306015.sp, bin.size=5000, num.cores=11)
WM306015.sp = makeBinary(WM306015.sp, mat="bmat");

### Genes.gr and promoters.gr
ov = findOverlaps(WM306015.sp@feature, promoter.gr);
idy = queryHits(ov);

promoter_ratio = SnapATAC::rowSums(WM306015.sp[,idy, mat="bmat"], mat="bmat") / SnapATAC::rowSums(WM306015.sp, mat="bmat");
plot(log(SnapATAC::rowSums(WM306015.sp, mat="bmat") + 1,10), promoter_ratio, cex=0.5, col="grey", xlab="log(UMI)", ylab="FIP Ratio", ylim=c(0,1 ));
idx = which(promoter_ratio > 0.2 & promoter_ratio < 0.8);
WM306015.sp = WM306015.sp[idx,]
WM306015.sp@metaData$MTRatio <- (WM306015.sp@metaData$CM+1) / (WM306015.sp@metaData$UQ+1)
WM306015.sp@metaData$PromoterRatio <- promoter_ratio[idx]

## After filteration
summarySnap(WM306015.sp)
plotBarcode(
  obj=WM306015.sp, 
	pdf.file.name=NULL, 
	pdf.width=7, 
	pdf.height=7, 
	col="grey",
	border="grey",
	breaks=50
	)













WM306016.sp = createSnap(
	file="WM306016.snap",
	sample="WM306016",
	num.cores=11
	)
### Weird prefiltering
plotBarcode(
  obj=WM306016.sp, 
	pdf.file.name=NULL, 
	pdf.width=7, 
	pdf.height=7, 
	col="grey",
	border="grey",
	breaks=50
	)

# Filter Cells
WM306016.sp = filterCells(
	obj=WM306016.sp,
	subset.names=c("fragment.num", "UMI"),
	low.thresholds=c(1000,500),
	high.thresholds=c(Inf, Inf)
	)

# ## Filter based upon previous filter
# trueBarcodes <- match(barcodes[,1], WM306016.sp@barcode)
# trueBarcodes <- na.omit(trueBarcodes)
# WM306016.sp <- WM306016.sp[trueBarcodes,, mat="bmat"]

## add bMAT
WM306016.sp = addBmatToSnap(WM306016.sp, bin.size=5000, num.cores=11)
WM306016.sp = makeBinary(WM306016.sp, mat="bmat");

### Genes.gr and promoters.gr
ov = findOverlaps(WM306016.sp@feature, promoter.gr);
idy = queryHits(ov);

promoter_ratio = SnapATAC::rowSums(WM306016.sp[,idy, mat="bmat"], mat="bmat") / SnapATAC::rowSums(WM306016.sp, mat="bmat");
plot(log(SnapATAC::rowSums(WM306016.sp, mat="bmat") + 1,10), promoter_ratio, cex=0.5, col="grey", xlab="log(UMI)", ylab="FIP Ratio", ylim=c(0,1 ));
idx = which(promoter_ratio > 0.2 & promoter_ratio < 0.8);
WM306016.sp = WM306016.sp[idx,]
WM306016.sp@metaData$MTRatio <- (WM306016.sp@metaData$CM+1) / (WM306016.sp@metaData$UQ+1)
WM306016.sp@metaData$PromoterRatio <- promoter_ratio[idx]

## After filteration
summarySnap(WM306016.sp)
plotBarcode(
  obj=WM306016.sp, 
	pdf.file.name=NULL, 
	pdf.width=7, 
	pdf.height=7, 
	col="grey",
	border="grey",
	breaks=50
	)

Sample_190827.sp = createSnap(
	file="190827_NS500556_0335_AHT3CLAFXY.Ece.snap",
	sample="190827",
	num.cores=11
	)
### Weird prefiltering
plotBarcode(
  obj=Sample_190827.sp, 
	pdf.file.name=NULL, 
	pdf.width=7, 
	pdf.height=7, 
	col="grey",
	border="grey",
	breaks=50
	)

# Filter Cells
Sample_190827.sp = filterCells(
	obj=Sample_190827.sp,
	subset.names=c("fragment.num", "UMI"),
	low.thresholds=c(1000,500),
	high.thresholds=c(Inf, Inf)
	)

# ## Filter based upon previous filter
# trueBarcodes <- match(barcodes[,1], Sample_190827.sp@barcode)
# trueBarcodes <- na.omit(trueBarcodes)
# Sample_190827.sp <- Sample_190827.sp[trueBarcodes,, mat="bmat"]

## add bMAT
Sample_190827.sp = addBmatToSnap(Sample_190827.sp, bin.size=5000, num.cores=11)
Sample_190827.sp = makeBinary(Sample_190827.sp, mat="bmat");

### Genes.gr and promoters.gr
ov = findOverlaps(Sample_190827.sp@feature, promoter.gr);
idy = queryHits(ov);

promoter_ratio = SnapATAC::rowSums(Sample_190827.sp[,idy, mat="bmat"], mat="bmat") / SnapATAC::rowSums(Sample_190827.sp, mat="bmat");
plot(log(SnapATAC::rowSums(Sample_190827.sp, mat="bmat") + 1,10), promoter_ratio, cex=0.5, col="grey", xlab="log(UMI)", ylab="FIP Ratio", ylim=c(0,1 ));
idx = which(promoter_ratio > 0.2 & promoter_ratio < 0.8);
Sample_190827.sp = Sample_190827.sp[idx,]
Sample_190827.sp@metaData$MTRatio <- (Sample_190827.sp@metaData$CM+1) / (Sample_190827.sp@metaData$UQ+1)
Sample_190827.sp@metaData$PromoterRatio <- promoter_ratio[idx]

## After filteration
summarySnap(Sample_190827.sp)
plotBarcode(
  obj=Sample_190827.sp, 
	pdf.file.name=NULL, 
	pdf.width=7, 
	pdf.height=7, 
	col="grey",
	border="grey",
	breaks=50
	)
summarySnap(Sample_190827.sp)
Test <- function(obj){
	if(nrow(obj@metaData) == 0){stop("metaData is empty")}
	barcode = obj@metaData;
	TotalNumberOfBarcodes <-length(obj@barcode)
	message("Median number of sequencing fragments: ", median(barcode$TN));
	message("Median number of uniquely mapped fragments: ", median(barcode$UQ));
	message("Median number of mappability ratio: ", round(median((barcode$UM+1)/(barcode$TN+1)),2));
	message("Median number of properly paired ratio: ", round(median((barcode$PP+1)/(barcode$UM+1)),2));
	message("Median number of duplicate ratio: ", round(median(1 - (barcode$UQ+1)/(barcode$PP+1)),2));
	message("Median number of chrM ratio: ", round(median((barcode$CM+1) / (barcode$UQ+1)),2));
	message("Median number of unique molecules (UMI): ", median(barcode$UQ));
}
Test(Sample_190827.sp)

snapNames <- ls(pattern = ".*.sp")
snapObjectList<- mget(snapNames)
snapObjectList <- snapObjectList[which(unlist(lapply(snapObjectList, nrow)) > 100)]


## All
bin.shared = Reduce(intersect, lapply(snapObjectList, function(x.sp) x.sp@feature$name));
snapObjectList <- lapply(snapObjectList, function(x.sp){
	idy = match(bin.shared, x.sp@feature$name);
	x.sp[,idy, mat="bmat"];
	})
AllSamplesDecember.sp = Reduce(snapRbind, snapObjectList);


library(GenomicRanges)
black_list <- read.table("Duke_Hg19SignalRepeatArtifactRegions.bed.gz")
black_list.gr = GRanges(
  black_list[,1],
  IRanges(black_list[,2], black_list[,3])
);
idy1 = queryHits(findOverlaps(AllSamplesDecember.sp@feature, black_list.gr))
idy2 = grep("chrM|random", AllSamplesDecember.sp@feature);
idy = unique(c(idy1, idy2));
AllSamplesDecember.sp = AllSamplesDecember.sp[,-idy, mat="bmat"]
library(tidyverse)
AllSamplesDecember.sp@metaData$Sample <- AllSamplesDecember.sp@sample






ggplot(AllSamplesDecember.lda.sp@metaData, 
       aes(x = factor(SampleNamePaper, levels=names(sort(table(AllSamplesDecember.lda.sp@metaData$SampleNamePaper), decreasing = T))), fill = Sample)) + 
  geom_bar() +
  theme_bw() +
  xlab("Sample") + 
  theme(axis.text.x = element_text(angle = 90)) + 
  ggtitle("Cells Per Sample") +  theme(legend.position = "none") 


  ### cisTopic
library(lda)
fixInNamespace(runLDA, "SnapATAC")
system.time({
	AllSamplesDecember.lda.sp = SnapATAC::runLDA(
		obj=AllSamplesDecember.sp, 
		input.mat = "bmat", 
		topic = c(10, 20, 30, 40, 50),
		method = "Z-score", 
		num.cores = 5, 
		min.cell = 10,
		seed.use = 10, 
		iterations = 500, 
		burnin = 250, 
		alpha = 50,
		alphaByTopic = TRUE, 
		beta = 0.1
		);
	})
saveRDS(AllSamplesDecember.lda.sp, "AllSamplesDecember.lda_March11_wGenes.sp.RDS")
```


``` {r,}
AllSamplesDecember.lda.sp <- readRDS("AllPlusDecemberMarch.RDS")
table(AllSamplesDecember.lda.sp@sample)

peakMatrix <- AllSamplesDecember.lda.sp@pmat
rownames(peakMatrix) <- AllSamplesDecember.lda.sp@barcode
head(peakMatrix)
# saveRDS(peakMatrix, "~/Box Sync/Alex_Data/Ece_scATAC/snapATAC_March/Figures_April/peakMatrix.RDS")
# saveRDS(AllSamplesDecember.lda.sp@metaData, "~/Box Sync/Alex_Data/Ece_scATAC/snapATAC_March/Figures_April/metaData.RDS")
# saveRDS(AllSamplesDecember.lda.sp@feature, "~/Box Sync/Alex_Data/Ece_scATAC/snapATAC_March/Figures_April/features.RDS")





plotDimReductPW(
    obj=AllSamplesDecember.lda.sp, 
    eigs.dims =seq(AllSamplesDecember.lda.sp@smat@sdev),
    point.size=0.3,
    point.color="grey",
    point.shape=19,
    point.alpha=0.6,
    down.sample=5000,
    pdf.file.name=NULL, 
    pdf.height=7, 
    pdf.width=7
    );

## KNN
AllSamplesDecember.lda.sp = runKNN(
    obj=AllSamplesDecember.lda.sp,
    eigs.dims =seq(AllSamplesDecember.lda.sp@smat@sdev),
    weight.by.lambda = T,
    k=15
    );

## Clustering


AllSamplesDecember.lda.sp = runCluster(
	obj=AllSamplesDecember.lda.sp,
	tmp.folder=tempdir(),
	louvain.lib="R-igraph",
	seed.use=10,
	resolution=1
	);

## Visualization
AllSamplesDecember.lda.sp = runViz(
	obj=AllSamplesDecember.lda.sp, 
	tmp.folder=tempdir(),
	dims=2,
	eigs.dims =seq(AllSamplesDecember.lda.sp@smat@sdev), 
	weight.by.sd=TRUE,
	method="umap",
	fast_tsne_path=NULL,
	Y.init=AllSamplesDecember.lda.sp@umap,
	seed.use=10,
	num.cores=5
	);

## Plot
plotViz(
	obj=AllSamplesDecember.lda.sp, 
	method="umap", 
	point.size=0.2, 
	point.shape=19, 
	point.alpha=0.8, 
  	point.color=AllSamplesDecember.lda.sp@cluster, 
	text.add=TRUE,
	text.size=0.5,
	text.color="black",
	text.halo.add=TRUE,
	text.halo.color="white",
	text.halo.width=0.2,
	down.sample=10000,
	pdf.file.name=NULL,
	pdf.width=7, 
	pdf.height=7,
	legend.add=T, 
	legend.pos = "top",
	main="snapATAC integrated LDA (cisTopic snapATAC)"
	)
library(SnapATAC)
barcodes <- read.table("190827_barcodes.txt", sep = "\t", header = T)
Is190827 <- grep("190827", AllSamplesDecember.lda.sp@sample)
ID54773 <- grep(paste(barcodes$ID54773, collapse = "|"), AllSamplesDecember.lda.sp@barcode)
ID54773 <- intersect(ID54773, Is190827)
ID56893 <- grep(paste(barcodes$ID56893, collapse = "|"), AllSamplesDecember.lda.sp@barcode)
ID56893 <- intersect(ID56893, Is190827)
AllSamplesDecember.lda.sp@sample[ID54773] <- "ID54773"
AllSamplesDecember.lda.sp@sample[ID56893] <- "ID56893"


plotViz(
	obj=AllSamplesDecember.lda.sp, 
	method="umap", 
	point.size=0.2, 
	point.shape=19, 
	point.alpha=0.8, 
  	point.color=AllSamplesDecember.lda.sp@metaData$, 
	text.add=TRUE,
	text.size=0.5,
	text.color="black",
	text.halo.add=TRUE,
	text.halo.color="white",
	text.halo.width=0.2,
	down.sample=10000,
	pdf.file.name=NULL,
	pdf.width=7, 
	pdf.height=7,
	legend.add=T, 
	legend.pos = "top",
	main="snapATAC integrated LDA (cisTopic snapATAC)"
	)

## Enriched gleason score
AllSamplesDecember.lda.sp@metaData$EnrichedCluster <- "G3"
AllSamplesDecember.lda.sp@metaData$EnrichedCluster[grep("^2$|^3$|^13$|^15$",AllSamplesDecember.lda.sp@cluster)] <- "G4"
AllSamplesDecember.lda.sp@metaData$EnrichedCluster[grep("^7$",AllSamplesDecember.lda.sp@cluster)] <- "Stroma"
AllSamplesDecember.lda.sp@metaData$EnrichedCluster[grep("^12$|^14",AllSamplesDecember.lda.sp@cluster)] <- "Immune"

## Soft Gleason Score
AllSamplesDecember.lda.sp@metaData$SoftGleasonScore <- "G3"
AllSamplesDecember.lda.sp@metaData$SoftGleasonScore[grep("^4\\+4$|^4\\+5$",AllSamplesDecember.lda.sp@metaData$GleasonScore)] <- "G4"
AllSamplesDecember.lda.sp@metaData$SoftGleasonScore <- "G3"
AllSamplesDecember.lda.sp@metaData$SoftGleasonScore[grep("^4\\+4$|^4\\+5$",AllSamplesDecember.lda.sp@metaData$GleasonScore)] <- "G4"

##IncreasedResolutionContrasts
AllSamplesDecember.lda.sp@metaData$IncreasedEnrichedResolution <- "G3"
AllSamplesDecember.lda.sp@metaData$IncreasedEnrichedResolution[grep("^13$|^15",AllSamplesDecember.lda.sp@cluster)] <- "G4_13_15"
AllSamplesDecember.lda.sp@metaData$IncreasedEnrichedResolution[grep("^3$",AllSamplesDecember.lda.sp@cluster)] <- "G4_3"
AllSamplesDecember.lda.sp@metaData$IncreasedEnrichedResolution[grep("^2$",AllSamplesDecember.lda.sp@cluster)] <- "G4_2"
AllSamplesDecember.lda.sp@metaData$IncreasedEnrichedResolution[which(AllSamplesDecember.lda.sp@cluster == "7" & AllSamplesDecember.lda.sp@metaData$SoftGleasonScore == "G4")] <- "Stroma_G4"
AllSamplesDecember.lda.sp@metaData$IncreasedEnrichedResolution[which(AllSamplesDecember.lda.sp@cluster == "7" & AllSamplesDecember.lda.sp@metaData$SoftGleasonScore == "G3")] <- "Stroma_G3"

AllSamplesDecember.lda.sp@metaData$IncreasedEnrichedResolution[which(AllSamplesDecember.lda.sp@metaData$EnrichedCluster == "Immune" & AllSamplesDecember.lda.sp@metaData$SoftGleasonScore == "G4")] <- "Immune_G4"

AllSamplesDecember.lda.sp@metaData$IncreasedEnrichedResolution[which(AllSamplesDecember.lda.sp@metaData$EnrichedCluster == "Immune" & AllSamplesDecember.lda.sp@metaData$SoftGleasonScore == "G3")] <- "Immune_G3"



library(tidyverse)
# SampleNames <- paste0("Sample_", 1:length(unique(AllSamplesDecember.lda.sp@sample)))
# names(SampleNames) <- unique(AllSamplesDecember.lda.sp@sample)
# AllSamplesDecember.lda.sp@metaData$SampleNamePaper <- plyr::revalue(AllSamplesDecember.lda.sp@sample, SampleNames)
# SampleNamesOut <- data.frame(CisTopic = names(SampleNames),
#               PaperName = SampleNames, 
#               row.names = 1:length(SampleNames))
# write.table(SampleNamesOut, 
#             file = "~/Box Sync/Alex_Data/Ece_scATAC/SnapATAC_February/SampleNameOut.txt", 
#             quote = F, 
#             sep = "\t", row.names = F, col.names = T)
plotUMAP <- function(snapObject, 
                     colorBy, 
                     colorByLegendName, 
                     LegendRow =2,
                     outFolder,
                     outFile) {
  plotDF <- data.frame(UMAP1=-1*snapObject@umap[,1],
             UMAP2=-1*snapObject@umap[,2],
             ColorBy=colorBy
             )
  colnames(plotDF)[3] <- colorByLegendName
  
  p1 <- ggplot(plotDF, aes_string(x = "UMAP1", y = "UMAP2", color = colorByLegendName)) +
    geom_point(size = 0.1) + 
    theme_bw() +
    theme(legend.position="bottom") +
    guides(color=guide_legend(override.aes = list(size=1),
                             nrow=LegendRow, 
                             byrow=TRUE)) +
    theme(axis.line=element_blank(),axis.text.x=element_blank(),
          axis.text.y=element_blank(),axis.ticks=element_blank(),
          axis.title.x=element_blank(),
          axis.title.y=element_blank(),
          panel.background=element_blank(),panel.border=element_blank(),panel.grid.major=element_blank(),
          panel.grid.minor=element_blank(),plot.background=element_blank())
  png(paste0(outFolder, "/", outFile, ".png"), 
      width = 11, 
      height = 8.5, 
      res = 300,
      units = "in")
  print(p1)
  dev.off()
  pdf(paste0(outFolder, "/", outFile, ".pdf"), 
      width = 11, 
      height = 8.5)
  print(p1)
  dev.off()
}
dev.off()
# Gleason
GleasonPattern <- read.csv("GleasonPattern.txt", sep = "\t", header=T)
## T Stage 
TStage <- as.character(GleasonPattern$T.stage)
names(TStage) <- GleasonPattern$Sample
AllSamplesDecember.lda.sp@metaData$TStage <- plyr::revalue(as.character(AllSamplesDecember.lda.sp@sample),
                                                                 TStage)

# CAPRA
CAPRAPattern <- read.csv("CAPRA.csv", sep = ",", header=T)
CAPRAPattern$Sample <- gsub("G44","G4",  CAPRAPattern$Sample)

## T Stage 
CAPRA <- as.character(CAPRAPattern$CAPRA.score)
names(CAPRA) <- CAPRAPattern$Sample
table(AllSamplesDecember.lda.sp@sample)
table(CAPRAPattern)
AllSamplesDecember.lda.sp@metaData$CAPRA <- plyr::revalue(as.character(AllSamplesDecember.lda.sp@sample),
                                                                 CAPRA)
table(AllSamplesDecember.lda.sp@metaData$CAPRA)
plotUMAP(snapObject = AllSamplesDecember.lda.sp, 
        colorBy = AllSamplesDecember.lda.sp@cluster,
        colorByLegendName = "Cluster",
        outFolder = "Figures_March/", 
        outFile = "UMAP_clusters")
plotUMAP(snapObject = AllSamplesDecember.lda.sp, 
        colorBy = AllSamplesDecember.lda.sp@metaData$GleasonScore,
        colorByLegendName = "GleasonScore",
        outFolder = "Figures_March/", 
        outFile = "UMAP_GleasonScore")
plotUMAP(snapObject = AllSamplesDecember.lda.sp, 
        colorBy = AllSamplesDecember.lda.sp@metaData$SoftGleasonScore,
        colorByLegendName = "SoftGleasonScore",
        outFolder = "Figures_March/", 
        outFile = "UMAP_SoftGleasonScore")
plotUMAP(snapObject = AllSamplesDecember.lda.sp,
        colorBy = AllSamplesDecember.lda.sp@metaData$TStage,
        colorByLegendName = "TScore",
        outFolder = "Figures_March/",
        outFile = "UMAP_TScore")
plotUMAP(snapObject = AllSamplesDecember.lda.sp,
        colorBy = AllSamplesDecember.lda.sp@metaData$CAPRA,
        colorByLegendName = "CAPRA",
        outFolder = "Figures_March/",
        outFile = "UMAP_CAPRA")
plotUMAP(snapObject = AllSamplesDecember.lda.sp, 
        colorBy = AllSamplesDecember.lda.sp@metaData$EnrichedCluster,
        colorByLegendName = "EnrichedGleasonScore",
        outFolder = "Figures_March/", 
        outFile = "UMAP_EnrichedGleasonScore")
plotUMAP(snapObject = AllSamplesDecember.lda.sp, 
        colorBy = AllSamplesDecember.lda.sp@metaData$SampleNamePaper,
        colorByLegendName = "Sample",
        outFolder = "Figures_March/", 
        outFile = "UMAP_Samplee")
plotUMAP(snapObject = AllSamplesDecember.lda.sp, 
        colorBy = AllSamplesDecember.lda.sp@metaData$IncreasedEnrichedResolution,
        colorByLegendName = "IncreasedResolution",
        outFolder = "Figures_March/", 
        outFile = "UMAP_IncreasedResolution")

## Heatmap 
library(pheatmap)
library(RColorBrewer)
gg_color_hue <- function(n) {
  hues = seq(15, 375, length = n + 1)
  hcl(h = hues, l = 65, c = 100)[1:n]
}
GetTopics <- function(model, Object) {
  modelMat <- t(scale(model$document_expects, center=TRUE, scale=TRUE))
  rownames(modelMat) <- paste(1:length(Object@metaData$barcode), Object@metaData$barcode, sep = "_")
  colnames(modelMat) <- paste("Topic", 1:ncol(modelMat), sep="_")
  return(modelMat)
}
heatmap <- function(Object, Topics, Anno_Vector, Anno_Name, outFolder, outFile) {
  anno_col <- data.frame(row.names = paste(1:length(Object@barcode), Object@barcode, sep = "_"),
                         Column=Anno_Vector)
  colnames(anno_col) <- Anno_Name
  num_colors = length(unique(anno_col[,1]))
  anno_colors = gg_color_hue(num_colors)
  names(anno_colors) <- sort(unique(anno_col[,1]))
  anno_colors <- list(Cluster = anno_colors)
  png(paste0(outFolder, "/", outFile, ".png"), 
      width = 11, 
      height = 8.5, 
      res = 300,
      units = "in")
  p1 <- pheatmap(data.matrix(Topics[order(anno_col[[Anno_Name]]),]),
                 hclustfun = function(x) hclust(x, method="ward.D2"),
                 scale = "row",
                 cluster_cols = T,
                 cluster_rows = F,show_rownames = F,
                 col=colorRampPalette(rev(brewer.pal(11, "RdBu"))[c(1:4,8:11)])(256),
                 annotation_row = anno_col,
                 annotation_colors = anno_colors,
                 cex=1, 
                 main =  paste0("Topic Per Cell By ", Anno_Name))
  print(p1)
  dev.off()
  pdf(paste0(outFolder, "/", outFile, ".pdf"), 
      width = 11, 
      height = 8.5)
  p1 <- pheatmap(data.matrix(Topics[order(anno_col[[Anno_Name]]),]),
                 hclustfun = function(x) hclust(x, method="ward.D2"),
                 scale = "row",
                 cluster_cols = T,
                 cluster_rows = F,show_rownames = F,
                 col=colorRampPalette(rev(brewer.pal(11, "RdBu"))[c(1:4,8:11)])(256),
                 annotation_row = anno_col,
                 annotation_colors = anno_colors,
                 cex=1, 
                 main = paste0("Topic Per Cell By ", Anno_Name))
  print(p1)
  dev.off()
}
selectedModel<- readRDS("selected.model_March.RDS")
heatmap(Object = AllSamplesDecember.lda.sp,
        Topics = GetTopics(selectedModel, AllSamplesDecember.lda.sp),
        Anno_Vector = AllSamplesDecember.lda.sp@cluster,
        Anno_Name = "Cluster",
        outFolder = "Figures_March/",
        outFile = "TopicHeatmap_clusters")
heatmap(Object = AllSamplesDecember.lda.sp, 
        Topics = GetTopics(selectedModel, AllSamplesDecember.lda.sp),
        Anno_Vector = AllSamplesDecember.lda.sp@metaData$GleasonScore, 
        Anno_Name = "Gleason Score",
        outFolder = "Figures_March/", 
        outFile = "TopicHeatmap_GleasonScore")
heatmap(Object = AllSamplesDecember.lda.sp, 
        Topics = GetTopics(selectedModel, AllSamplesDecember.lda.sp),
        Anno_Vector = AllSamplesDecember.lda.sp@metaData$EnrichedCluster, 
        Anno_Name = "Enriched Cluster",
        outFolder = "Figures_March/", 
        outFile = "TopicHeatmap_Enrichedclusters")
heatmap(Object = AllSamplesDecember.lda.sp, 
        Topics = GetTopics(selectedModel, AllSamplesDecember.lda.sp),
        Anno_Vector = AllSamplesDecember.lda.sp@metaData$SampleNamePaper, 
        Anno_Name = "Sample Name",
        outFolder = "Figures_March/", 
        outFile = "TopicHeatmap_SampleName")
heatmap(Object = AllSamplesDecember.lda.sp, 
        Topics = GetTopics(selectedModel, AllSamplesDecember.lda.sp),
        Anno_Vector = AllSamplesDecember.lda.sp@metaData$SoftGleasonScore, 
        Anno_Name = "Soft Gleason Score",
        outFolder = "Figures_March/", 
        outFile = "TopicHeatmap_SoftGleasonScore")
heatmap(Object = AllSamplesDecember.lda.sp, 
        Topics = GetTopics(selectedModel, AllSamplesDecember.lda.sp),
        Anno_Vector = AllSamplesDecember.lda.sp@cluster, 
        Anno_Name = "Cluster",
        outFolder = "Figures_March/", 
        outFile = "TopicHeatmap_clusters")




getTopicsBedFiles <- function(TopicMatrix, 
                              snapObject, 
                              outFolder, 
                              min.cell = 10,
                              nRegions = 1000) {
  if (!dir.exists(outFolder)) {
    dir.create(outFolder)
  }
  ## Get feature matrix output 
  FeatureTopicMatrix <- t(TopicMatrix)
  
  ## Filter out cells not used in the LDA tranformation
  idx <- which(Matrix::colSums(snapObject@bmat) > min.cell)
  for (i in 1:ncol(FeatureTopicMatrix)) {
    topic.gr <- snapObject@feature[idx][order(FeatureTopicMatrix[,i], decreasing = T)[1:nRegions]]
    convertGRangesToBED(GR = topic.gr, outBedFile = paste0(outFolder, "/topic", as.character(i), ".bed"))
  }
  return(idx)
}
getTopicWordMatrix <- function(selectedModel) {
  topic.mat <- selectedModel$topics
  ## cisTopic code  
  beta  <- 0.1
  modelMat <-  (topic.mat + beta)/rowSums(topic.mat + beta)
  return(modelMat)
}
library(GenomicRanges)
getTopicsBedFiles(TopicMatrix = getTopicWordMatrix(selectedModel), 
                  snapObject = AllSamplesDecember.lda.sp, 
                  outFolder = "/AllPlusDecemberTopics_March/",
                  nRegions = 1500)





AllSamplesDecember.lda@met


AllSamplesDecember.lda.G3.G4Only.sp@metaData$SampleNamePaper <- plyr::revalue(AllSamplesDecember.lda.G3.G4Only.sp@sample, SampleNames)
plotUMAP(snapObject = AllSamplesDecember.lda.G3.G4Only.sp, 
        colorBy = AllSamplesDecember.lda.G3.G4Only.sp@cluster,s
        colorByLegendName = "Cluster")
plotUMAP(snapObject = AllSamplesDecember.lda.G3.G4Only.sp, 
        colorBy = AllSamplesDecember.lda.G3.G4Only.sp@metaData$GleasonScore,
        colorByLegendName = "GleasonScore")
plotUMAP(snapObject = AllSamplesDecember.lda.G3.G4Only.sp, 
        colorBy = AllSamplesDecember.lda.G3.G4Only.sp@metaData$SoftGleasonScore,
        colorByLegendName = "SoftGleasonScore")
plotUMAP(snapObject = AllSamplesDecember.lda.G3.G4Only.sp, 
        colorBy = AllSamplesDecember.lda.G3.G4Only.sp@metaData$TStage,
        colorByLegendName = "TScore")
plotUMAP(snapObject = AllSamplesDecember.lda.G3.G4Only.sp, 
        colorBy = AllSamplesDecember.lda.G3.G4Only.sp@metaData$EnrichedCluster,
        colorByLegendName = "EnrichedGleasonScore")
plotUMAP(snapObject = AllSamplesDecember.lda.G3.G4Only.sp, 
        colorBy = AllSamplesDecember.lda.G3.G4Only.sp@metaData$SampleNamePaper,
        colorByLegendName = "Sample", LegendRow = 3)

AllSamplesDecember.lda.sp[]
findOverlaps(genes.sel.gr["ERG"]


GeneFunction_Distrubution <- function(snapObject, 
                                      Gene, 
                                      OutFolder = "Figures_March/GenePlot_Distrubution/") {
  geneMetaTable <- data.frame(Sample = snapObject@metaData$SampleNamePaper, 
                            Gene = snapObject@gmat[,Gene])
  colnames(geneMetaTable)[2] <- Gene
  outFile <- paste(Gene, "Distrubution_plot_BySample.pdf", sep = "_")
  pdf(file = paste0(OutFolder, "/", outFile), width = 11, height = 8.5)
  p1 <- ggplot(geneMetaTable, aes_string(x = "Sample", y= Gene, color = "Sample")) +
    geom_violin(scale = "width", draw_quantiles = c(0.25, 0.5, 0.75)) +
    geom_jitter(size = 0.1) +
    coord_flip() + 
    theme_minimal() + 
    theme(legend.position="bottom") +
    guides(color=guide_legend(override.aes = list(size=1),
                              nrow=3, 
                              byrow=TRUE)) + 
    ggtitle(paste(Gene, "in Gene Matrix By Sample"))
  print(p1)
  dev.off()
  return(p1)
}
# GeneFunction_Distrubution <- function(snapObject, 
#                                       Gene, 
#                                       OutFolder = "~/Box Sync/Alex_Data/Ece_scATAC/SnapATAC_February/Figures_March/GenePlot_Distrubution/") {
#   geneMetaTable <- data.frame(Sample = snapObject@metaData$SampleNamePaper, 
#                             Gene = snapObject@gmat[,Gene])
#   colnames(geneMetaTable)[2] <- Gene
#   return(geneMetaTable)
# }
# 
# 
# colnames(AllSamplesDecember.lda.sp@gmat)[colSums(AllSamplesDecember.lda.sp@gmat) == 0]
AllSamplesDecember.lda.sp@bmat
GeneFunction_Distrubution(AllSamplesDecember.lda.sp, "ERG")
GeneFunction_Distrubution(AllSamplesDecember.lda.sp, "AR")
GeneFunction_Distrubution(AllSamplesDecember.lda.sp, "GAPDH")
GeneFunction_Distrubution(AllSamplesDecember.lda.sp, "ETV1")
GeneFunction_Distrubution(AllSamplesDecember.lda.sp, "ETV4")
GeneFunction_Distrubution(AllSamplesDecember.lda.sp, "FLI1")

GeneFunction_Distrubution_Pos_Neg <- function(snapObject, 
                                      Gene, 
                                      OutFolder = "Figures_March/GenePlot_Distrubution/",
                                      max.cutoff = NULL) {
  geneMetaTable <- data.frame(ERG_Status = factor(snapObject@metaData$ERGPositive), 
                              Gene = snapObject@gmat[,Gene])
  colnames(geneMetaTable)[2] <- Gene
  if (length(max.cutoff) > 0) {
    geneMetaTable <- geneMetaTable[geneMetaTable[[Gene]] < max.cutoff,]
  }
  outFile <- paste(Gene, "Distrubution_plot_ByERGStatus.pdf", sep = "_")
  pdf(file = paste0(OutFolder, "/", outFile), width = 11, height = 8.5)
  p1 <- ggplot(geneMetaTable, aes_string(x = "ERG_Status", y= Gene, color = "ERG_Status")) +
    geom_violin(scale = "width", draw_quantiles = c(0.25, 0.5, 0.75)) +
    geom_jitter(size= 0.1) +
    coord_flip() +
    theme_minimal() +
    theme(legend.position="bottom") +
    guides(color=guide_legend(override.aes = list(size=1),
                              nrow=1,
                              byrow=TRUE)) +
    ggtitle(paste(Gene, "in Gene Matrix By ERG Status")) 
  print(p1)
  dev.off()
  return(p1)
}
GeneFunction_Distrubution_Pos_Neg(AllSamplesDecember.lda.sp, "ERG")

dev.off()
test <- test[which(test$ERG_Status == "ERG_Pos"),]
test <- test[which(test$ERG < 2500),]
ggplot(test, aes(x = ERG_Status, y = ERG)) + geom_violin() 
plotViz(
	obj=AllSamplesDecember.lda.sp, 
	method="umap", 
	point.size=0.2, 
	point.shape=19, 
	point.alpha=0.8, 
  point.color=AllSamplesDecember.lda.sp@metaData$SampleNamePaper, 
	text.add=TRUE,
	text.size=0.5,
	text.color="black",
	text.halo.add=TRUE,
	text.halo.color="white",
	text.halo.width=0.2,
	down.sample=10000,
	pdf.file.name=NULL,
	pdf.width=7, 
	pdf.height=7,
	legend.add=T, 
	legend.pos = "top",
	main="snapATAC integrated LDA (cisTopic snapATAC)"
	)

AllSamplesDecember.lda.G3.G4Only.sp<- AllSamplesDecember.lda.sp[grep("Immune|Stroma", AllSamplesDecember.lda.sp@metaData$EnrichedCluster, invert = T),]

plotViz(
	obj=AllSamplesDecember.lda.G3.G4Only.sp, 
	method="umap", 
	point.size=0.2, 
	point.shape=19, 
	point.alpha=0.8, 
  point.color=AllSamplesDecember.lda.G3.G4Only.sp@cluster, 
	text.add=TRUE,
	text.size=0.5,
	text.color="black",
	text.halo.add=TRUE,
	text.halo.color="white",
	text.halo.width=0.2,
	down.sample=10000,
	pdf.file.name=NULL,
	pdf.width=7, 
	pdf.height=7,
	legend.add=T, 
	legend.pos = "top",
	main="snapATAC integrated LDA (cisTopic snapATAC)"
	)
plotUMAP(snapObject = AllSamplesDecember.lda.G3.G4Only.sp, 
        colorBy = AllSamplesDecember.lda.G3.G4Only.sp@cluster,
        colorByLegendName = "Cluster",
        outFolder = "Figures_March/", 
        outFile = "G3G4Only_UMAP_clusters")
plotUMAP(snapObject = AllSamplesDecember.lda.G3.G4Only.sp, 
        colorBy = AllSamplesDecember.lda.G3.G4Only.sp@metaData$GleasonScore,
        colorByLegendName = "GleasonScore",
        outFolder = "Figures_March/", 
        outFile = "G3G4Only_UMAP_GleasonScore")
plotUMAP(snapObject = AllSamplesDecember.lda.G3.G4Only.sp, 
        colorBy = AllSamplesDecember.lda.G3.G4Only.sp@metaData$SoftGleasonScore,
        colorByLegendName = "SoftGleasonScore",
        outFolder = "Figures_March/", 
        outFile = "G3G4Only_UMAP_SoftGleasonScore")
# plotUMAP(snapObject = AllSamplesDecember.lda.G3.G4Only.sp, 
#         colorBy = AllSamplesDecember.lda.G3.G4Only.sp@metaData$TStage,
#         colorByLegendName = "TScore",
#         outFolder = "Figures_March/", 
#         outFile = "G3G4Only_UMAP_TScore")
plotUMAP(snapObject = AllSamplesDecember.lda.G3.G4Only.sp, 
        colorBy = AllSamplesDecember.lda.G3.G4Only.sp@metaData$EnrichedCluster,
        colorByLegendName = "EnrichedGleasonScore",
        outFolder = "Figures_March/", 
        outFile = "G3G4Only_UMAP_EnrichedGleasonScore")
plotUMAP(snapObject = AllSamplesDecember.lda.G3.G4Only.sp, 
        colorBy = AllSamplesDecember.lda.G3.G4Only.sp@metaData$SampleNamePaper,
        colorByLegendName = "Sample",
        outFolder = "Figures_March/", 
        outFile = "G3G4Only_UMAP_Samplee")
plotUMAP(snapObject = AllSamplesDecember.lda.G3.G4Only.sp, 
        colorBy = AllSamplesDecember.lda.G3.G4Only.sp@metaData$IncreasedEnrichedResolution,
        colorByLegendName = "IncreasedResolution",
        outFolder = "Figures_March/", 
        outFile = "G3G4Only_UMAP_IncreasedResolution")


##################################
#### NEW #########################
##################################

table(AllSamplesDecember.lda.sp@metaData$EnrichedCluster)
##### MACS for EnrichedCluster
clusters.sel = names(table(AllSamplesDecember.lda.sp@metaData$EnrichedCluster))[which(table(AllSamplesDecember.lda.sp@metaData$EnrichedCluster) > 100)];

setwd("MACS_EnrichedCluster//")
library(parallel)
peaks.enriched.ls = mclapply(seq(clusters.sel), function(i){
    print(clusters.sel[i]);
    peaks = runMACS(
        obj=AllSamplesDecember.lda.sp[which(AllSamplesDecember.lda.sp@metaData$EnrichedCluster==clusters.sel[i]),], 
        output.prefix=paste0("AllSamplesDecember.lda.", gsub(" ", "_", clusters.sel)[i]),
        path.to.snaptools="miniconda3/bin/snaptools",
        path.to.macs="miniconda3/envs/scitools-env/bin/macs2",
        gsize="hs", # mm, hs, etc
        buffer.size=500, 
        num.cores=1,
        macs.options="--nomodel --shift 100 --ext 200 --qval 5e-2 -B --SPMR",
        tmp.folder=tempdir()
   );
	peaks
 	}, mc.cores=4);

## MACS Sample
samples.sel = names(table(AllSamplesDecember.lda.sp@metaData$SampleNamePaper))
setwd("MACS_/")
library(parallel)
peaks.sample.ls = mclapply(seq(samples.sel), function(i){
    print(samples.sel[i]);
    peaks = runMACS(
        obj=AllSamplesDecember.lda.sp[which(AllSamplesDecember.lda.sp@metaData$SampleNamePaper==samples.sel[i]),], 
        output.prefix=paste0("AllSamplesDecember.lda.", gsub(" ", "_", samples.sel)[i]),
        path.to.snaptools="miniconda3/bin/snaptools",
        path.to.macs="miniconda3/envs/scitools-env/bin/macs2",
        gsize="hs", # mm, hs, etc
        buffer.size=500, 
        num.cores=1,
        macs.options="--nomodel --shift 100 --ext 200 --qval 5e-2 -B --SPMR",
        tmp.folder=tempdir()
   );
	peaks
 	}, mc.cores=9);
peaks.sample.ls
clusters.sel = names(table(AllSamplesDecember.lda.sp@cluster))
clusters.sel
setwd("MACS_Cluster//")
library(parallel)
peaks.cluster.ls = mclapply(seq(clusters.sel), function(i){
    print(clusters.sel[i]);
    peaks = runMACS(
        obj=AllSamplesDecember.lda.sp[which(AllSamplesDecember.lda.sp@cluster==clusters.sel[i]),], 
        output.prefix=paste0("AllSamplesDecember.lda.", gsub(" ", "_", clusters.sel)[i]),
        path.to.snaptools="miniconda3/bin/snaptools",
        path.to.macs="miniconda3/envs/scitools-env/bin/macs2",
        gsize="hs", # mm, hs, etc
        buffer.size=500, 
        num.cores=1,
        macs.options="--nomodel --shift 100 --ext 200 --qval 5e-2 -B --SPMR",
        tmp.folder=tempdir()
   );
	peaks
 	}, mc.cores=9);
getwd()
library(GenomicRanges)
peaks.names = list.files(pattern = "narrow", path = c("/MACS_Sample/", 
                                                      "/MACS_Cluster/",
                                                      "/MACS_Cluster/"), full.names = T)
peaks.names = list.files(pattern = "narrow", path = c("/MACS_Sample/"), full.names = T)
peak.gr.ls = lapply(peaks.names, function(x){
    peak.df = read.table(x)
    GRanges(peak.df[,1], IRanges(peak.df[,2], peak.df[,3]))
  })
peak.gr = reduce(Reduce(c, peak.gr.ls));
seqlevels(peak.gr) <- gsub("b\'(.*)\'", "\\1", seqlevels(peak.gr))
peak.gr <- readRDS("peaks.all.gr")
peak.gr
#### Differential Accessiblity 
#################################
AllSamplesDecember.lda.sp <- readRDS("AllPlusDecemberMarch.RDS")
library(SnapATAC)
library(GenomicAlignments)
AllSamplesDecember.lda.sp <- createPmat(
  obj=AllSamplesDecember.lda.sp,
  peaks=peak.gr,
  ncell.chunk=20,
  do.par=TRUE,
  num.cores=11
)

collapseIdentForGrep <- function(stringV) {
  stringV <- paste(stringV, collapse = "$|^")
  stringV <- paste0("^", stringV)
  stringV <- paste0(stringV, "$")
  return(stringV)
}
#### DIFFERENTIAL EXPRESSION
differentialExpression <- function(snapObject, 
                                   DEFactor, 
                                   ident.1, 
                                   ident.1.Name = "Ident.1", 
                                   ident.2 = NULL, 
                                   ident.2.Name = "Ident.2",
                                   outFolder = "DARs//",
                                   peaks.gr = peak.gr,
                                   PValue = NULL,
                                   clusterNegMethod = FALSE,
                                   FDR = 5e-2,
                                   testMethod = "exactTest") {
  tmp.sp <- snapObject
  objectFile <- paste0(ident.1.Name, "VS", ident.2.Name)
  outFile <- paste0(outFolder, "/", objectFile)
  
  ## Set up contrasts
  tmp.DEFactor <- as.vector(DEFactor)
  ident.1 <- collapseIdentForGrep(ident.1)
  tmp.DEFactor[grep(ident.1, tmp.DEFactor)] <- ident.1.Name
  if (length(ident.2) > 0) {
    ## Change name of Ident 2 and Add to Factor
    ident.2 <- collapseIdentForGrep(ident.2)
    tmp.DEFactor[grep(ident.2, tmp.DEFactor)] <- ident.2.Name
    ## Only Keep Cells of Interest
    cellsOfInterest <- grep(paste(ident.1, ident.2, sep = "|"), as.vector(DEFactor))
    tmp.DEFactor <- tmp.DEFactor[cellsOfInterest]
    tmp.sp <- tmp.sp[cellsOfInterest,]
  } else {
    tmp.DEFactor[grep(ident.1, tmp.DEFactor, invert = T)] <- ident.2.Name
  }
  tmp.sp@cluster <- factor(tmp.DEFactor)
  if (clusterNegMethod == TRUE) {
    DARs <- findDAR(
      obj=tmp.sp,
      input.mat = "pmat",
      cluster.pos = ident.1.Name,
      cluster.neg=NULL,
      cluster.neg.method="knn",
      bcv = 0.4,
      test.method=testMethod,
      seed.use=10
    );
  } else {
    DARs <- findDAR(
      obj=tmp.sp,
      input.mat = "pmat",
      cluster.pos = ident.1.Name,
      cluster.neg = ident.2.Name,
      bcv = 0.4,
      test.method=testMethod,
      seed.use=10
    );
  }
  
  DARs$FDR = p.adjust(DARs$PValue, method="BH");
  DARs$Peak = paste0("Peak_",rownames(DARs))
  
  if (length(PValue) > 0) {
    idy = which(DARs$PValue < PValue & DARs$logFC > 0);
  } else {
    idy = which(DARs$FDR < FDR & DARs$logFC > 0);
  }
  numGenes <- length(idy)
  pdf(paste0(outFile, ".pdf"), width = 8, height = 6)
  p1 <- plot(DARs$logCPM, DARs$logFC, 
       pch=19, cex=0.1, col="grey", 
       ylab="logFC", xlab="logCPM",
       main=paste(ident.1.Name, "vs.", ident.2.Name, "( nRegions = ", numGenes, ")")
  )
  p1 <- points(DARs$logCPM[idy], 
         DARs$logFC[idy], 
         pch=19, 
         cex=0.5, 
         col="red"
  )
  p1 <- abline(h = 0, lwd=1, lty=2);
  dev.off()
  
  ##
  DARs <- DARs[idy,]
  DARs <- DARs[order(DARs$FDR),]
  assign(paste0("DARs.", objectFile), DARs, envir = .GlobalEnv)
  write.table(DARs, 
              file = paste0(outFile, ".txt"), 
              quote = F, sep = "\t", row.names = FALSE)
  
  ##
  Enriched.peaks <- peaks.gr[idy]
  convertGRangesToBED(GR = Enriched.peaks, outBedFile = paste0(outFile, ".bed"))
}
peak.gr
peak.gr$Peak <- paste0("Peak_", 1:length(peak.gr))
convertGRangesToBED <- function(GR, outBedFile) {
  outDF <- data.frame(seqnames=seqnames(GR),
                      starts=start(GR)-1,
                      ends=end(GR),
                      names = GR$Peak,
                      scores=c(rep("1000", length(GR))),
                      strands=c(rep(".", length(GR))))
  write.table(outDF, file= outBedFile, quote=F, sep="\t", row.names=F, col.names=F)
}
differentialExpression(snapObject = AllSamplesDecember.lda.sp, 
                       DEFactor = AllSamplesDecember.lda.sp@metaData$EnrichedCluster, 
                       ident.1 = c("G4"),
                       ident.1.Name = "G4", 
                       ident.2 = "G3", 
                       ident.2.Name = "G3")
differentialExpression(snapObject = AllSamplesDecember.lda.sp, 
                       DEFactor = AllSamplesDecember.lda.sp@metaData$EnrichedCluster, 
                       ident.1 = c("G4"),
                       ident.1.Name = "G4", 
                       ident.2 = "G3", 
                       ident.2.Name = "G3_ClusterNeg",
                       clusterNegMethod = T)
differentialExpression(snapObject = AllSamplesDecember.lda.sp, 
                       DEFactor = AllSamplesDecember.lda.sp@metaData$EnrichedCluster, 
                       ident.1 = c("G4"),
                       ident.1.Name = "G4", 
                       ident.2 = "G3", 
                       ident.2.Name = "G3_LRT",
                       testMethod = "LRT")
differentialExpression(snapObject = AllSamplesDecember.lda.sp, 
                       DEFactor = AllSamplesDecember.lda.sp@metaData$EnrichedCluster, 
                       ident.1 = c("G4"),
                       ident.1.Name = "G4", 
                       ident.2 = "G3", 
                       ident.2.Name = "G3_PValue",
                       PValue = 0.01)
differentialExpression(snapObject = AllSamplesDecember.lda.sp, 
                       DEFactor = AllSamplesDecember.lda.sp@metaData$EnrichedCluster, 
                       ident.1 = c("G3"),
                       ident.1.Name = "G3", 
                       ident.2 = "G4", 
                       ident.2.Name = "G4_PValue",
                       PValue = 0.01)
differentialExpression(snapObject = AllSamplesDecember.lda.sp, 
                       DEFactor = AllSamplesDecember.lda.sp@metaData$EnrichedCluster, 
                       ident.1 = c("G3"),
                       ident.1.Name = "G3", 
                       ident.2 = "G4", 
                       ident.2.Name = "G4")








differentialExpression(snapObject = AllSamplesDecember.lda.sp, 
                       DEFactor = AllSamplesDecember.lda.sp@cluster, 
                       ident.1 = c(8,10,4,11),
                       ident.1.Name = "TMPRSS2", 
                       ident.2 = c(5,1,9,6,16,2,3,13,15), 
                       ident.2.Name = "Others")








differentialExpression(snapObject = AllSamplesDecember.lda.sp, 
                       DEFactor = AllSamplesDecember.lda.sp@metaData$IncreasedEnrichedResolution, 
                       ident.1 = c("G4_2"),
                       ident.1.Name = "G4_2", 
                       ident.2 = c("G4_13_15", "G4_3"),
                       ident.2.Name = "G4_3_13_15")
differentialExpression(snapObject = AllSamplesDecember.lda.sp, 
                       DEFactor = AllSamplesDecember.lda.sp@metaData$IncreasedEnrichedResolution, 
                       ident.1 = c("G4_2"),
                       ident.1.Name = "G4_2", 
                       ident.2 = c("G4_13_15", "G4_3"),
                       ident.2.Name = "G4_3_13_15_PValue",
                       PValue = 0.01)
differentialExpression(snapObject = AllSamplesDecember.lda.sp, 
                       DEFactor = AllSamplesDecember.lda.sp@metaData$IncreasedEnrichedResolution, 
                       ident.1 = c("G4_3"),
                       ident.1.Name = "G4_3", 
                       ident.2 = c("G4_13_15", "G4_2"),
                       ident.2.Name = "G4_13_15_2")
differentialExpression(snapObject = AllSamplesDecember.lda.sp, 
                       DEFactor = AllSamplesDecember.lda.sp@metaData$IncreasedEnrichedResolution, 
                       ident.1 = c("G4_3"),
                       ident.1.Name = "G4_3", 
                       ident.2 = c("G4_13_15", "G4_2"),
                       ident.2.Name = "G4_13_15_2_PValue",
                       PValue = 0.01)
differentialExpression(snapObject = AllSamplesDecember.lda.sp, 
                       DEFactor = AllSamplesDecember.lda.sp@metaData$IncreasedEnrichedResolution, 
                       ident.1 = c("G4_13_15"),
                       ident.1.Name = "G4_13_15", 
                       ident.2 = c("G4_3", "G4_2"),
                       ident.2.Name = "G4_3_2")
differentialExpression(snapObject = AllSamplesDecember.lda.sp, 
                       DEFactor = AllSamplesDecember.lda.sp@metaData$IncreasedEnrichedResolution, 
                       ident.1 = c("G4_13_15"),
                       ident.1.Name = "G4_13_15", 
                       ident.2 = c("G4_3", "G4_2"),
                       ident.2.Name = "G4_3_2_PValue",
                       PValue = 0.01)
differentialExpression(snapObject = AllSamplesDecember.lda.sp, 
                       DEFactor = AllSamplesDecember.lda.sp@metaData$IncreasedEnrichedResolution, 
                       ident.1 = c("G4_2"),
                       ident.1.Name = "G4_2", 
                       ident.2 = "G3",
                       ident.2.Name = "G3")
differentialExpression(snapObject = AllSamplesDecember.lda.sp, 
                       DEFactor = AllSamplesDecember.lda.sp@metaData$IncreasedEnrichedResolution, 
                       ident.1 = c("G4_2"),
                       ident.1.Name = "G4_2", 
                       ident.2 = "G3",
                       ident.2.Name = "G3_PValue",
                       PValue = 0.01)
differentialExpression(snapObject = AllSamplesDecember.lda.sp, 
                       DEFactor = AllSamplesDecember.lda.sp@metaData$IncreasedEnrichedResolution, 
                       ident.1 = c("G4_13_15"),
                       ident.1.Name = "G4_13_15", 
                       ident.2 = "G3",
                       ident.2.Name = "G3")
differentialExpression(snapObject = AllSamplesDecember.lda.sp, 
                       DEFactor = AllSamplesDecember.lda.sp@metaData$IncreasedEnrichedResolution, 
                       ident.1 = c("G4_13_15"),
                       ident.1.Name = "G4_13_15", 
                       ident.2 = "G3",
                       ident.2.Name = "G3_PValue",
                       PValue = 0.01)
differentialExpression(snapObject = AllSamplesDecember.lda.sp, 
                       DEFactor = AllSamplesDecember.lda.sp@metaData$IncreasedEnrichedResolution, 
                       ident.1 = c("G4_3"),
                       ident.1.Name = "G4_3", 
                       ident.2 = "G3",
                       ident.2.Name = "G3")
differentialExpression(snapObject = AllSamplesDecember.lda.sp, 
                       DEFactor = AllSamplesDecember.lda.sp@metaData$IncreasedEnrichedResolution, 
                       ident.1 = c("G4_3"),
                       ident.1.Name = "G4_3", 
                       ident.2 = "G3",
                       ident.2.Name = "G3_PValue",
                       PValue = 0.01)
differentialExpression(snapObject = AllSamplesDecember.lda.sp, 
                       DEFactor = AllSamplesDecember.lda.sp@metaData$IncreasedEnrichedResolution, 
                       ident.1 = c("Immune_G4"),
                       ident.1.Name = "Immune_G4", 
                       ident.2 = "Immune_G3",
                       ident.2.Name = "Immune_G3")
differentialExpression(snapObject = AllSamplesDecember.lda.sp, 
                       DEFactor = AllSamplesDecember.lda.sp@metaData$IncreasedEnrichedResolution, 
                       ident.1 = c("Immune_G4"),
                       ident.1.Name = "Immune_G4", 
                       ident.2 = "Immune_G3",
                       ident.2.Name = "Immune_G3_PValue",
                       PValue = 0.01)
differentialExpression(snapObject = AllSamplesDecember.lda.sp, 
                       DEFactor = AllSamplesDecember.lda.sp@metaData$IncreasedEnrichedResolution, 
                       ident.1 = c("Immune_G3"),
                       ident.1.Name = "Immune_G3", 
                       ident.2 = "Immune_G4",
                       ident.2.Name = "Immune_G4_LRT",
                       testMethod = "LRT")
differentialExpression(snapObject = AllSamplesDecember.lda.sp, 
                       DEFactor = AllSamplesDecember.lda.sp@metaData$IncreasedEnrichedResolution, 
                       ident.1 = c("Immune_G3"),
                       ident.1.Name = "Immune_G3", 
                       ident.2 = "Immune_G4",
                       ident.2.Name = "Immune_G4_PValue",
                       PValue = 0.01)
differentialExpression(snapObject = AllSamplesDecember.lda.sp, 
                       DEFactor = AllSamplesDecember.lda.sp@metaData$IncreasedEnrichedResolution, 
                       ident.1 = c("Stroma_G3"),
                       ident.1.Name = "Stroma_G3", 
                       ident.2 = "Stroma_G4",
                       ident.2.Name = "Stroma_G4")
differentialExpression(snapObject = AllSamplesDecember.lda.sp, 
                       DEFactor = AllSamplesDecember.lda.sp@metaData$IncreasedEnrichedResolution, 
                       ident.1 = c("Stroma_G3"),
                       ident.1.Name = "Stroma_G3", 
                       ident.2 = "Stroma_G4",
                       ident.2.Name = "Stroma_G4_LRT",
                       testMethod = "LRT")
differentialExpression(snapObject = AllSamplesDecember.lda.sp, 
                       DEFactor = AllSamplesDecember.lda.sp@metaData$IncreasedEnrichedResolution, 
                       ident.1 = c("Stroma_G3"),
                       ident.1.Name = "Stroma_G3", 
                       ident.2 = "Stroma_G4",
                       ident.2.Name = "Stroma_G4_PValue", 
                       PValue = 0.01)
differentialExpression(snapObject = AllSamplesDecember.lda.sp, 
                       DEFactor = AllSamplesDecember.lda.sp@metaData$IncreasedEnrichedResolution, 
                       ident.1 = c("Stroma_G4"),
                       ident.1.Name = "Stroma_G4", 
                       ident.2 = "Stroma_G3",
                       ident.2.Name = "Stroma_G3_PValue", 
                       PValue = 0.01)


addTopicsToSnapObject <- function(model, Object) {
  modelMat <- t(scale(model$document_expects, center=TRUE, scale=TRUE))
  rownames(modelMat) <- paste(1:length(Object@barcode), Object@barcode, sep="_")
  colnames(modelMat) <- paste("Topic", 1:ncol(modelMat), sep="_")
  Object@metaData <- cbind(Object@metaData, modelMat)
  return(Object)
}

AllSamplesDecember.lda.sp <- addTopicsToSnapObject(selectedModel, AllSamplesDecember.lda.sp)
for (i in 1:30) {
  topic <- paste0("Topic_", i)
  pdf(paste0("TopicUMAP//", topic, ".pdf"),
      width = 11,
      height = 8.5)
  plotFeatureSingle(
    obj=AllSamplesDecember.lda.sp,
    feature.value=AllSamplesDecember.lda.sp@metaData[,topic],
    method="umap", 
    main=gsub("_", "", topic),
    point.size=0.1, 
    point.shape=19, 
    down.sample=10000,
    quantiles=c(0, 0.9)
  )
  dev.off()
  png(paste0("TopicUMAP//", topic, ".png"),
      width = 11,
      height = 8.5, 
      res = 300, 
      units = 'in')
  plotFeatureSingle(
    obj=AllSamplesDecember.lda.sp,
    feature.value=AllSamplesDecember.lda.sp@metaData[,topic],
    method="umap", 
    main=gsub("_", "", topic),
    point.size=0.1, 
    point.shape=19, 
    down.sample=10000,
    quantiles=c(0, 0.9)
  )
  dev.off()
  
}












table(AllSamplesDecember.lda.sp@metaData$EnrichedCluster)
sum(AllSamplesDecember.lda.sp@pmat[AllSamplesDecember.lda.sp@metaData$EnrichedCluster == "G4",])
sum(AllSamplesDecember.lda.sp@pmat[AllSamplesDecember.lda.sp@metaData$EnrichedCluster == "G3",])

#################################################################
#####################      G3 Only        #######################
#################################################################


g3Only.lda.sp<- AllSamplesDecember.lda.sp[AllSamplesDecember.lda.sp@metaData$EnrichedCluster == "G3",]
g3Only.lda.sp
system.time({
	G3Only.lda.sp = SnapATAC::runLDA(
		obj=g3Only.lda.sp, 
		input.mat = "bmat", 
		topic = c(10, 20, 30, 40, 50),
		method = "Z-score", 
		num.cores = 5, 
		min.cell = 10,
		seed.use = 10, 
		iterations = 500, 
		burnin = 250, 
		alpha = 50,
		alphaByTopic = TRUE, 
		beta = 0.1
		);
	})
G3Only.lda.sp <- readRDS("G3only.rds")

G3Only.lda.sp = runKNN(
    obj=G3Only.lda.sp,
    eigs.dims =seq(G3Only.lda.sp@smat@sdev),
    weight.by.lambda = T,
    k=15
    );

## Clustering
library(leiden);
G3Only.lda.sp = runCluster(
	obj=G3Only.lda.sp,
	tmp.folder=tempdir(),
	louvain.lib="R-igraph",
	seed.use=10,
	resolution=0.3
	);

## Visualization
G3Only.lda.sp = runViz(
	obj=G3Only.lda.sp, 
	tmp.folder=tempdir(),
	dims=2,
	eigs.dims =seq(G3Only.lda.sp@smat@sdev), 
	weight.by.sd=TRUE,
	method="umap",
	fast_tsne_path=NULL,
	Y.init=G3Only.lda.sp@umap,
	seed.use=10,
	num.cores=10
	);

plotViz(
	obj=G3Only.lda.sp, 
	method="umap", 
	point.size=0.2, 
	point.shape=19, 
	point.alpha=0.8, 
  point.color=G3Only.lda.sp@cluster, 
	text.add=TRUE,
	text.size=0.5,
	text.color="black",
	text.halo.add=TRUE,
	text.halo.color="white",
	text.halo.width=0.2,
	down.sample=10000,
	pdf.file.name=NULL,
	pdf.width=7, 
	pdf.height=7,
	legend.add=T, 
	legend.pos = "top",
	main="snapATAC integrated LDA (cisTopic snapATAC)"
	)

plotUMAP(snapObject = G3Only.lda.sp, 
        colorBy = G3Only.lda.sp@cluster,
        colorByLegendName = "Cluster",
        outFolder = "Figures_March/", 
        outFile = "G3OnlyCistopicRerun_UMAP_Cluster")
plotUMAP(snapObject = G3Only.lda.sp, 
        colorBy = G3Only.lda.sp@metaData$SampleNamePaper,
        colorByLegendName = "SampleName",
        outFolder = "Figures_March/", 
        outFile = "G3OnlyCistopicRerun_UMAP_SampleName")

## Plots 
peaks2granges <- function(peakstable) {
  temp.gr <- with(peakstable, GRanges(chr, IRanges(start, end), strand, name, score, signalValue, pValue, qValue, peaks))
  return(temp.gr)
}
loadMACS <- function(infiles, outfiles) {
  for (i in 1:length(infiles)) {
    temp.peaks <- read.table(infiles[i], sep="\t", header=FALSE)
    colnames(temp.peaks) <- c("chr", "start", "end", "name", "score", "strand", "signalValue", "pValue", "qValue", "peaks")
    temp.peaks$strand <- "*"
    assign(outfiles[i], temp.peaks, envir = .GlobalEnv)
    temp.peaks.gr <- peaks2granges(temp.peaks)
    outnames.gr <- paste0(outfiles, ".gr")
    assign(outnames.gr[i], temp.peaks.gr, envir = .GlobalEnv)
  }
}
files <- list.files(path = c("MACS_Sample/", "MACS_EnrichedCluster/"),
                    pattern=".*peaks_correct.narrowPeak$",
                    recursive = TRUE,
                    full.names = T)

### Fix Filenames for Output in R
filenames <- gsub("_peaks_correct.narrowPeak", "", files)
filenames <- gsub(".*December.lda.", "", filenames)
filenames
library(GenomicRanges)
loadMACS(files, filenames)

library(ChIPseeker)
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
txdb <- TxDb.Hsapiens.UCSC.hg19.knownGene
TCGA.gr <- TCGA
granges <- ls(pattern = ".*.gr$")
granges <- granges[grep("190827.gr", granges, invert = T)]

table <- data.frame()
for (gr in granges) {
  print(gr)
  peakAnno <- annotatePeak(get(gr), tssRegion=c(-3000, 3000),
                         TxDb=txdb, annoDb="org.Hs.eg.db")
  peakAnnotation <- peakAnno@annoStat
  peakAnnotation$Sample <- gr
  table <- rbind(table, peakAnnotation)
}
# table
# table$Sample[grep("peaks.lda.gr", table$Sample)] <- "All_Samples_MACS"
# table$Sample <- gsub(".gr$", "", table$Sample)
library(tidyverse)

ggplot(table, aes(x=Sample, y=Frequency, fill=Feature)) + 
    geom_bar(position = "fill",stat = "identity")+
    scale_y_continuous(labels = scales::percent_format(), expand = c(0,0)) +
    theme_classic() +
    theme(axis.text.x = element_text(angle = 90, hjust = 1), 
          axis.ticks.x = element_blank(),
          axis.line.x=element_blank()) +
    xlab("Sample") +
    ylab("Percentage") +
  ggtitle("Distrubtion of Peak Annotation")
# table[table$Sample == "TCGA",]
# table(table$Sample)
# spread(table, key = Feature, value = Frequency)


library(TxDb.Hsapiens.UCSC.hg38.knownGene)
txdb <- TxDb.Hsapiens.UCSC.hg38.knownGene
granges <- "TCGA.gr"

for (gr in granges) {
  print(gr)
  peakAnno <- annotatePeak(get(gr), tssRegion=c(-3000, 3000),
                         TxDb=txdb, annoDb="org.Hs.eg.db")
  peakAnnotation <- peakAnno@annoStat
  peakAnnotation$Sample <- gr
  table <- rbind(table, peakAnnotation)
}
table$Sample <- gsub(".gr$", "", table$Sample)
table$Sample <- gsub("peak", "AllSamples_sciATAC", table$Sample)
table$Sample <- factor(table$Sample, levels = c("TCGA",  "AllSamples_sciATAC", "G3", "G4", "Immune", "Stroma", "Sample_1", "Sample_10", "Sample_11", "Sample_12", "Sample_13", "Sample_14", "Sample_15", "Sample_16", "Sample_17", "Sample_18", "Sample_2", "Sample_3",  "Sample_4",  "Sample_5",  "Sample_6","Sample_7",  "Sample_8" , "Sample_9"))
table$Sample
tidyverse::ggplot(table %>% filter(Sample == "TCGA" | Sample == "AllSamples_sciATAC" |Sample == "G4" | Sample == "G3"), aes(x=Sample, y=Frequency, fill=Feature)) + 
    geom_bar(position = "fill",stat = "identity")+
    scale_y_continuous(labels = scales::percent_format(), expand = c(0,0)) +
    theme_classic() +
    theme(axis.text.x = element_text(angle = 90, hjust = 1), 
          axis.ticks.x = element_blank(),
          axis.line.x=element_blank()) +
    xlab("Sample") +
    ylab("Percentage") +
  ggtitle("Distrubtion of Peak Annotation")


granges <- "peak.gr"
for (gr in granges) {
  print(gr)
  peakAnno <- annotatePeak(get(gr), tssRegion=c(-3000, 3000),
                           TxDb=txdb, annoDb="org.Hs.eg.db")
  peakAnnotation <- peakAnno@annoStat
  peakAnnotation$Sample <- gr
  table <- rbind(table, peakAnnotation)
}
table1 <- table %>% filter(Sample != "AllSamples_sciATAC")
table1$Sample <- gsub(".gr$", "", table1$Sample)
table1$Sample <- gsub("peak", "AllSamples_sciATAC", table1$Sample)
table1$Sample <- factor(table1$Sample, levels = c("TCGA",  "AllSamples_sciATAC", "G3", "G4", "Immune", "Stroma", "Sample_1", "Sample_10", "Sample_11", "Sample_12", "Sample_13", "Sample_14", "Sample_15", "Sample_16", "Sample_17", "Sample_18", "Sample_2", "Sample_3",  "Sample_4",  "Sample_5",  "Sample_6","Sample_7",  "Sample_8" , "Sample_9"))
"Sample_3",  "Sample_4",  "Sample_5",  "Sample_6","Sample_7",  "Sample_8" , "Sample_9"))
ggplot(table1 %>% filter(Sample == "TCGA" | Sample == "AllSamples_sciATAC" |Sample == "G4" | Sample == "G3"), aes(x=Sample, y=Frequency, fill=Feature)) + 
  geom_bar(position = "fill",stat = "identity")+
  scale_y_continuous(labels = scales::percent_format(), expand = c(0,0)) +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1), 
        axis.ticks.x = element_blank(),
        axis.line.x=element_blank()) +
  xlab("Sample") +
  ylab("Percentage") +
  ggtitle("Distrubtion of Peak Annotation")




# calculate the ensemble signals for each cluster
ensemble.ls = lapply(split(seq(length(AllSamplesDecember.lda.sp@cluster)), AllSamplesDecember.lda.sp@cluster), function(x){
	SnapATAC::colMeans(AllSamplesDecember.lda.sp[x,], mat="bmat");
	})
# cluster using 1-cor as distance  
hc = hclust(as.dist(1 - cor(t(do.call(rbind, ensemble.ls)))), method="ward.D2");
plot(hc, hang=-1, xlab="")


## Cluster
SampleCluster <- data.frame(Sample = AllSamplesDecember.lda.sp@metaData$SampleNamePaper, Cluster = AllSamplesDecember.lda.sp@cluster)
ggplot(SampleCluster, aes(x= Cluster, fill = Cluster)) + geom_bar(position = "dodge") + facet_wrap(~Sample)


OntologyFiles <- list.files(pattern = ".*tsv$",
                            path = "GO_GREAT_tsv/", 
                            full.names = T, recursive = T)
OntologyOutPDF <- gsub(".tsv$", "_smallBars.pdf", OntologyFiles)
OntologyOutPNG <- gsub(".tsv$", "_smallBars.png", OntologyFiles)
OntologyFiles
for (i in 1:length(OntologyOutPDF)) {
  OntTable <- read.table(OntologyFiles[i], sep = "\t", stringsAsFactors = F, header = F)
  OntTable <- OntTable[,c(1,3)]
  colnames(OntTable) <- c("TermName", "Binom.Raw.P.Value")
  paths <- unlist(strsplit(OntologyFiles[i], "/"))
  Title <- gsub(".tsv", "", paths[length(paths)])
  OntTable$Binom.log.P.Value <- signif(-log10(OntTable$Binom.Raw.P.Value), 4)
  p1 <- ggplot(OntTable, aes(x = reorder(TermName,Binom.log.P.Value), y = Binom.log.P.Value)) + 
    geom_bar(stat = "identity", fill = "#66b3ff") + 
    coord_flip() +
    theme_minimal() +
    ylab("-log10(Binomial p value)") + 
    xlab("Term Name") +
    ggtitle(Title) +
    theme(axis.text=element_text(size=15))
  pdf(OntologyOutPDF[i], width = 11, height = 3.5)
  print(p1)
  dev.off()
  png(OntologyOutPNG[i], width = 11, height = 3.5,res = 300, units = "in")
  print(p1)
  dev.off()
}





OntologyFiles <- list.files(pattern = ".*tsv$",
                            path = "SuppFig15/",
                            full.names = T, 
                            recursive = T)
OntologyOutPDF <- gsub(".tsv$", "_smallBars.pdf", OntologyFiles)
OntologyOutPNG <- gsub(".tsv$", "_smallBars.png", OntologyFiles)
OntologyFiles
for (i in 1:length(OntologyOutPDF)) {
  OntTable <- read.table(OntologyFiles[i], sep = "\t", stringsAsFactors = F, header = F)
  OntTable <- OntTable[,c(1,3)]
  colnames(OntTable) <- c("TermName", "Binom.Raw.P.Value")
  paths <- unlist(strsplit(OntologyFiles[i], "/"))
  Title <- gsub(".tsv", "", paths[length(paths)])
  OntTable$Binom.log.P.Value <- signif(-log10(OntTable$Binom.Raw.P.Value), 4)
  p1 <- ggplot(OntTable, aes(x = reorder(TermName,Binom.log.P.Value), y = Binom.log.P.Value)) + 
    geom_bar(stat = "identity", fill = "#66b3ff") + 
    coord_flip() +
    theme_minimal() +
    ylab("-log10(Binomial p value)") + 
    xlab("Term Name") +
    ggtitle(Title) +
    theme(axis.text=element_text(size=15))
  pdf(OntologyOutPDF[i], width = 11, height = 3.5)
  print(p1)
  dev.off()
  png(OntologyOutPNG[i], width = 11, height = 3.5,res = 300, units = "in")
  print(p1)
  dev.off()
}

OntologyOutPDF <- gsub(".tsv$", "_mediumBars.pdf", OntologyFiles)
OntologyOutPNG <- gsub(".tsv$", "_mediumBars.png", OntologyFiles)
for (i in 1:length(OntologyOutPDF)) {
  OntTable <- read.table(OntologyFiles[i], sep = "\t", stringsAsFactors = F, header = F)
  OntTable <- OntTable[,c(1,3)]
  colnames(OntTable) <- c("TermName", "Binom.Raw.P.Value")
  paths <- unlist(strsplit(OntologyFiles[i], "/"))
  Title <- gsub(".tsv", "", paths[length(paths)])
  OntTable$Binom.log.P.Value <- signif(-log10(OntTable$Binom.Raw.P.Value), 4)
  p1 <- ggplot(OntTable, aes(x = reorder(TermName,Binom.log.P.Value), y = Binom.log.P.Value)) + 
    geom_bar(stat = "identity", fill = "#66b3ff") + 
    coord_flip() +
    theme_minimal() +
    ylab("-log10(Binomial p value)") + 
    xlab("Term Name") +
    ggtitle(Title) +
    theme(axis.text=element_text(size=10))
  pdf(OntologyOutPDF[i], width = 11, height = 3.5)
  print(p1)
  dev.off()
  png(OntologyOutPNG[i], width = 11, height = 3.5,res = 300, units = "in")
  print(p1)
  dev.off()
}


OntologyOutPDF <- gsub(".tsv$", "_largeBars.pdf", OntologyFiles)
OntologyOutPNG <- gsub(".tsv$", "_largeBars.png", OntologyFiles)
for (i in 1:length(OntologyOutPDF)) {
  OntTable <- read.table(OntologyFiles[i], sep = "\t", stringsAsFactors = F, header = F)
  OntTable <- OntTable[,c(1,3)]
  colnames(OntTable) <- c("TermName", "Binom.Raw.P.Value")
  paths <- unlist(strsplit(OntologyFiles[i], "/"))
  Title <- gsub(".tsv", "", paths[length(paths)])
  OntTable$Binom.log.P.Value <- signif(-log10(OntTable$Binom.Raw.P.Value), 4)
  p1 <- ggplot(OntTable, aes(x = reorder(TermName,Binom.log.P.Value), y = Binom.log.P.Value)) + 
    geom_bar(stat = "identity", fill = "#66b3ff") + 
    coord_flip() +
    theme_minimal() +
    ylab("-log10(Binomial p value)") + 
    xlab("Term Name") +
    ggtitle(Title) +
    theme(axis.text=element_text(size=7))
  pdf(OntologyOutPDF[i], width = 11, height = 3.5)
  print(p1)
  dev.off()
  png(OntologyOutPNG[i], width = 11, height = 3.5,res = 300, units = "in")
  print(p1)
  dev.off()
}







OntologyFiles <- list.files(pattern = ".*tsv$",
                            path = "SuppFig9/",
                            full.names = T, 
                            recursive = T)
OntologyOutPDF <- gsub(".tsv$", "_smallBars.pdf", OntologyFiles)
OntologyOutPNG <- gsub(".tsv$", "_smallBars.png", OntologyFiles)
OntologyFiles
for (i in 1:length(OntologyOutPDF)) {
  OntTable <- read.table(OntologyFiles[i], sep = "\t", stringsAsFactors = F, header = F)
  OntTable <- OntTable[,c(1,3)]
  colnames(OntTable) <- c("TermName", "Binom.Raw.P.Value")
  paths <- unlist(strsplit(OntologyFiles[i], "/"))
  Title <- gsub(".tsv", "", paths[length(paths)])
  OntTable$Binom.log.P.Value <- signif(-log10(OntTable$Binom.Raw.P.Value), 4)
  p1 <- ggplot(OntTable, aes(x = reorder(TermName,Binom.log.P.Value), y = Binom.log.P.Value)) + 
    geom_bar(stat = "identity", fill = "#66b3ff") + 
    coord_flip() +
    theme_minimal() +
    ylab("-log10(Binomial p value)") + 
    xlab("Term Name") +
    ggtitle(Title) +
    theme(axis.text=element_text(size=15))
  pdf(OntologyOutPDF[i], width = 11, height = 3.5)
  print(p1)
  dev.off()
  png(OntologyOutPNG[i], width = 11, height = 3.5,res = 300, units = "in")
  print(p1)
  dev.off()
}

OntologyOutPDF <- gsub(".tsv$", "_mediumBars.pdf", OntologyFiles)
OntologyOutPNG <- gsub(".tsv$", "_mediumBars.png", OntologyFiles)
for (i in 1:length(OntologyOutPDF)) {
  OntTable <- read.table(OntologyFiles[i], sep = "\t", stringsAsFactors = F, header = F)
  OntTable <- OntTable[,c(1,3)]
  colnames(OntTable) <- c("TermName", "Binom.Raw.P.Value")
  paths <- unlist(strsplit(OntologyFiles[i], "/"))
  Title <- gsub(".tsv", "", paths[length(paths)])
  OntTable$Binom.log.P.Value <- signif(-log10(OntTable$Binom.Raw.P.Value), 4)
  p1 <- ggplot(OntTable, aes(x = reorder(TermName,Binom.log.P.Value), y = Binom.log.P.Value)) + 
    geom_bar(stat = "identity", fill = "#66b3ff") + 
    coord_flip() +
    theme_minimal() +
    ylab("-log10(Binomial p value)") + 
    xlab("Term Name") +
    ggtitle(Title) +
    theme(axis.text=element_text(size=10))
  pdf(OntologyOutPDF[i], width = 11, height = 3.5)
  print(p1)
  dev.off()
  png(OntologyOutPNG[i], width = 11, height = 3.5,res = 300, units = "in")
  print(p1)
  dev.off()
}


OntologyOutPDF <- gsub(".tsv$", "_largeBars.pdf", OntologyFiles)
OntologyOutPNG <- gsub(".tsv$", "_largeBars.png", OntologyFiles)
for (i in 1:length(OntologyOutPDF)) {
  OntTable <- read.table(OntologyFiles[i], sep = "\t", stringsAsFactors = F, header = F)
  OntTable <- OntTable[,c(1,3)]
  colnames(OntTable) <- c("TermName", "Binom.Raw.P.Value")
  paths <- unlist(strsplit(OntologyFiles[i], "/"))
  Title <- gsub(".tsv", "", paths[length(paths)])
  OntTable$Binom.log.P.Value <- signif(-log10(OntTable$Binom.Raw.P.Value), 4)
  p1 <- ggplot(OntTable, aes(x = reorder(TermName,Binom.log.P.Value), y = Binom.log.P.Value)) + 
    geom_bar(stat = "identity", fill = "#66b3ff") + 
    coord_flip() +
    theme_minimal() +
    ylab("-log10(Binomial p value)") + 
    xlab("Term Name") +
    ggtitle(Title) +
    theme(axis.text=element_text(size=7))
  pdf(OntologyOutPDF[i], width = 11, height = 3.5)
  print(p1)
  dev.off()
  png(OntologyOutPNG[i], width = 11, height = 3.5,res = 300, units = "in")
  print(p1)
  dev.off()
}


```



